---
title: "AS in LUNG TIL: examples"
author: "Jason Li"
date: "`r Sys.Date()`"
# output:
#   html_notebook:
#     fig_caption: yes
#     toc: yes
#   html_document:
#     df_print: paged
#     toc: yes
---

```{r, echo=FALSE}
# !diagnostics off
knitr::opts_chunk$set(
  comment = "#>",
  collapse = TRUE,
  echo = FALSE,
  warning = FALSE,
  message = FALSE,
  error = FALSE
)
```

# Setting parameters 
```{r}
wd = getwd()
source(paste0(wd, "/utils/ToolKit.R"))
source(paste0(wd, "/utils/utils.R"))
target_CDTypes = c("CD8", "CD4") # MAIT will be removed when running CD8+ T cells
# out_dir = "/your/path/"
input_path = paste0(wd, "/../IDEA_data/01.IDEAinputs/scLUNGscescet.RDS") 
out_dir = "/data2/jason/data/IDEA_data/02.IDEAoutputs/"
path_to_GENEINFO = paste0(wd, "/../IDEA_data/01.IDEAinputs/GRCh38.p12.biomart.transcripts.txt")

usage_cutoff = 0.75
transcript_express_cutoff = 10
step = 100
cores_to_use = 10
```


# Loading 
```{r input}
wd = getwd()
source(paste0(wd, "/utils/ToolKit.R"))
source(paste0(wd, "/utils/utils.R"))
# ---- loading dependencies ----
suppressPackageStartupMessages(library("tidyverse"))
suppressPackageStartupMessages(library("reshape2"))
suppressPackageStartupMessages(library("parallel"))
suppressPackageStartupMessages(library("ggforce"))
suppressPackageStartupMessages(library("patchwork"))
suppressPackageStartupMessages(library("openxlsx"))
theme_set(theme_classic(base_size = 20))


# load data ----
target_CDType = "CD8"
in_dir = paste0(wd, "/../IDEA_data/02.IDEAoutputs/")

load(paste0(in_dir, "02.", target_CDType, "_analysis_bundle.RDa"))

all_obj_lst = read_rds(
            path = paste0(in_dir, "/", target_CDType, "_enrichment_obj_lst.rds"))
```

# To check the mean usage level of a given gene
```{r, fig.width=8, fig.height=6}
target_gene = "PTPRC"
t_usageMat = usageMat[usageMat %>% rownames() %>% grep(pattern = target_gene, value = T), ]
t_usageMat_tbl = t_usageMat %>% melt(value.name = "usage", varnames = c("transcript_id", "cell_id")) %>% mutate(cluster = pCells$characteristics..majorCluster[match(cell_id, pCells$title)])

t_usageMat_tbl %>% group_by(transcript_id, cluster) %>%
  summarise(mean = mean(usage), n = n(), sd = sd(usage), se = sd / sqrt(n) ) %>%
  mutate(edge = ifelse(transcript_id %in% c("PTPRC-201", "PTPRC-209"), "canonical", "non-canonical")) %>% 
  ggplot(aes( x = cluster, y = mean, color = edge, fill = transcript_id)) +
  geom_bar(stat = "identity", position = position_dodge()) +
  theme(axis.text.x = element_text(angle = 90, hjust = 1)) +
  scale_color_manual(values = c("canonical" = "black", "non-canonical" = "white")) +
  labs(y = "Mean usage level",
       x = "",
       color = "Isoform",
       fill = "Isoform")

```

# To show structure of transcripts of interest
```{r}
txsForStructures = c("CD8A-201", "CD8A-202", "CD8A-203")
test_struct_tbl = STRUCT.INFO %>% dplyr::filter(`Transcript name` %in% txsForStructures) %>% dplyr::mutate(bioType = type2bio(GENE.INFO$`Transcript type`[match(`Transcript name`, GENE.INFO$`Transcript name`)]))
local_drawTranscriptStrucutre(test_struct_tbl)

annotation_tsvs = paste0(tsv_path, txsForStructures, ".fasta.tsv.txt")
t_domain_info_tbl = local_loadDomains(annotation_tsvs)
if (nrow(t_domain_info_tbl)) {
  t_domain_info_tbl = t_domain_info_tbl %>% dplyr::mutate(entryName = `Transcript name`)
  p5 = local_drawProteinStructure(t_domain_info_tbl)
  p5 = p5 + theme(
    legend.position = "bottom",
    legend.direction = "vertical",
    axis.line.x = element_blank(),
    axis.ticks.x = element_blank(),
    axis.text.x = element_blank()
  ) + labs(x = "")
} else{
  p5 = ggplot()
}
p5

```