---
title: "AS in LUNG TIL: calculate Enrichment"
author: "Jason Li"
date: "`r Sys.Date()`"
# output:
#   html_notebook:
#     fig_caption: yes
#     toc: yes
#   html_document:
#     df_print: paged
#     toc: yes
---

```{r, echo=FALSE}
# !diagnostics off
knitr::opts_chunk$set(
  comment = "#>",
  collapse = TRUE,
  echo = FALSE,
  warning = FALSE,
  message = FALSE,
  error = FALSE
)
```

# Loading
```{r input}
wd = getwd()
source(paste0(wd, "/utils/ToolKit.R"))
source(paste0(wd, "/utils/utils.R"))

# ---- loading dependencies ----
suppressPackageStartupMessages(library("tidyverse"))
suppressPackageStartupMessages(library("reshape2"))
suppressPackageStartupMessages(library("parallel"))
suppressPackageStartupMessages(library("ggforce"))
suppressPackageStartupMessages(library("patchwork"))
# suppressPackageStartupMessages(library("multidplyr"))
# suppressPackageStartupMessages(library("pheatmap"))
suppressPackageStartupMessages(library("openxlsx"))
# library(extrafont)
# suppressPackageStartupMessages(library("scater"))
theme_set(theme_classic(base_size = 20))


# load data ----
sceObj = read_rds(paste0(wd, "/../IDEA_data/06.data_totry/p01.scLUNGsceset.RDS_2018July18")) 

# reduce to only CD8 cells for easier calculation
# chn_clusters = sceObj$`characteristics: majorCluster` %>% unique() %>% grep(pattern = "CD8", value = T) %>% grep(pattern = "C7", invert = T, value = T)
chn_clusters = sceObj$characteristics..majorCluster %>% unique() %>% grep(pattern = "CD8", value = T) %>% grep(pattern = "C7", invert = T, value = T)
# chn_cells = sceObj$`characteristics: majorCluster` %in% chn_clusters
chn_cells = sceObj$characteristics..majorCluster %in% chn_clusters
sceObj = sceObj[, chn_cells]

# remove txs not in the normal annotation
wd = getwd()
GENE.INFO = read_tsv(paste0(wd, "/../IDEA_data/02.IDEAinputs/GRCh38.p12.biomart.transcripts.txt"))
GENE.INFO = GENE.INFO %>% dplyr::filter(`Chromosome/scaffold name` %in% c(as.character(1:22), "X", "Y", "MT"))

keep_tx = rownames(sceObj) %in% GENE.INFO$`Transcript name`
loginfo("Filter out txs on alternative haplotypes: ", sum(keep_tx), "/", length(keep_tx), " remained.")

sceObj = sceObj[keep_tx, ]

rm(list = grep(ls(), pattern = "^chn_", value = T))
invisible(gc())

```

```{r all expression distribution}
pCells = as_tibble(colData(sceObj))
exprMat_all = assay(sceObj,"tpm")
exprMat_sum = rowsum(exprMat_all, group = rownames(exprMat_all) %>% trans2gene())
exprMat_sum_tbl = exprMat_sum %>% melt(varnames = c("Gene_name", "Cell_id"), value.name = "tpm") %>% dplyr::mutate(cluster = pCells$characteristics..majorCluster[match(Cell_id, pCells$title)], tissue = pCells$characteristics..tissueType[match(Cell_id, pCells$title)]) %>% as_tibble()

# t_idx = sample(1:nrow(exprMat_all), round(0.2 * nrow(exprMat_all), digits = 0))
# t_cell_idx = sample(1:ncol(exprMat_all), round(0.2 * ncol(exprMat_all), digits = 0))
# t_exprMat_tbl = log10(exprMat_all[t_idx, t_cell_idx] + 1) %>% melt(varnames = c("id", "cell_id"), value.name = "log10tpm") %>% as_tibble() %>% dplyr::mutate(cluster = pCells$characteristics..majorCluster[match(cell_id, pCells$title)]) %>% as_tibble()
# 
# require(ggridges)
# ggplot(t_exprMat_tbl %>% dplyr::filter(log10tpm > 0)) + 
#   ggridges::geom_density_ridges2(aes(x = log10tpm, y = cluster, fill = cluster))+
#   geom_vline(xintercept = 1, linetype = 2) +
#   theme(legend.position = "bottom")

# t_idx = sample(1:nrow(usageMat), round(0.2 * nrow(usageMat), digits = 0))
# t_cell_idx = sample(1:ncol(usageMat), round(0.2 * ncol(usageMat), digits = 0))
# t_usageMat_tbl = usageMat[t_idx, t_cell_idx] %>% melt(varnames = c("id", "cell_id"), value.name = "usage") %>% as_tibble() %>% dplyr::mutate(cluster = pCells$characteristics..majorCluster[match(cell_id, pCells$title)]) %>% as_tibble()

# require(ggridges)
# ggplot(t_usageMat_tbl %>% dplyr::filter(usage > 0)) + 
#   ggridges::geom_density_ridges2(aes(x = usage, y = cluster, fill = cluster))+
#   scale_x_continuous(breaks = c(0,0.25,0.5,0.75,1)) +
#   labs(y = "Cell density")
# 
# rm(list = grep(ls(), pattern = "^t_", value = T))
# invisible(gc())
```

```{r data preparation}
# genes with more than 1 annotation ----
genes_anntx_counts = GENE.INFO %>% group_by(`Gene name`) %>% summarise(ann_count = length(unique(`Transcript name`)))
genes_morethan1ann = (genes_anntx_counts %>% dplyr::filter(ann_count > 1))$`Gene name`
genes_morethan1ann = intersect(genes_morethan1ann, rownames(exprMat_sum))

# loading ----
usageMat = assay(sceObj, "usage")
sceObj = sceObj[, !is.na(usageMat[1, ])]
pGenes = as_tibble(rowData(sceObj))
pCells = as_tibble(colData(sceObj))

# extract usages and tpms ----
usageMat = assay(sceObj, "usage")
exprMat = assay(sceObj,"tpm")


# remove transcripts with only one txs annotate ----
keep_tx = (rownames(usageMat) %>% trans2gene()) %in% genes_morethan1ann
usageMat = usageMat[keep_tx, ]
exprMat = exprMat[keep_tx, ]

# remove transcripts with only one txs remained ----
genes_morethan1remain = rownames(usageMat) %>% trans2gene() %>% table()
genes_morethan1remain = names(genes_morethan1remain[genes_morethan1remain>1])
keep_tx = (rownames(usageMat) %>% trans2gene()) %in% genes_morethan1ann
usageMat = usageMat[keep_tx, ]
exprMat = exprMat[keep_tx, ]

# cleanup ----
loginfo(length(genes_morethan1ann), "/", nrow(exprMat_sum), " genes with more than 1 annotations")
loginfo(length(genes_morethan1remain), "/", nrow(exprMat_sum), " genes with more than 1 remian")

loginfo("Filter out isoforms: ", nrow(usageMat), "/", nrow(pGenes))

rm(sceObj, keep_tx)
invisible(gc())
```

```{r}
# # t_expr_markers = exprMat > transcript_express_cutoff
# t_usageMat = usageMat
# # t_usageMat[!t_expr_markers] = 0
# t_dom_counts = apply(t_usageMat, 1, function(x) sum(x > usage_cutoff, na.rm = T))
# t_dom_counts_tbl = tibble(
#   txs_id = names(t_dom_counts),
#   domn_cell_count = t_dom_counts
# ) %>% dplyr::mutate(gene_name = trans2gene(txs_id))
# 
# 
# t_gene_dmn_iso_count_tbl = t_dom_counts_tbl %>% group_by(gene_name) %>% summarise(dmn_iso_count = sum(domn_cell_count != 0), all_iso_count = length(domn_cell_count)) %>% dplyr::mutate(dmn_ratio = dmn_iso_count/all_iso_count)
# 
# t_gene_dmn_iso_count_tbl %>% ggplot() + geom_histogram(aes(x = dmn_ratio)) + scale_y_log10()
# t_gene_dmn_iso_count_tbl %>% ggplot() + geom_histogram(aes(x = dmn_iso_count)) + scale_y_log10()
# 
# # how much of those with one dominant isoforms are because of only one tx annotated
# t_annotation_counts = (GENE.INFO %>% dplyr::filter(`Gene name` %in% (t_gene_dmn_iso_count_tbl %>% dplyr::filter(dmn_iso_count ==1))$gene_name))$`Gene name` %>% table()

```

```{r load trajectory}
# CD4_trajectory_tbl = bind_cols(CD4_trajectory_coord, MajorType = pCells$characteristics..majorCluster[match(CD4_trajectory_coord$Cell, pCells$title)])
# CD8_trajectory_tbl = bind_cols(CD8_trajectory_coord, MajorType = pCells$characteristics..majorCluster[match(CD8_trajectory_coord$Cell, pCells$title)])
# ggplot(CD8_trajectory_tbl, aes (x = Component1, y = Component2, color = MajorType)) +
#   geom_point()

```


```{r functions definition}
usage_cutoff = 0.75
transcript_express_cutoff = 10
# function from p03.counts2fisher.R
counts2FISHER_p = function(countMat, alternative = "greater"){
  sum_trans = apply(countMat, 1, sum)
  sum_cat = apply(countMat, 2, sum)
  sum_all = sum(countMat)
  
  minor_trans = sum_trans - countMat
  minor_cat = t(apply(countMat, 1, function(x) sum_cat - x))
  minor_all = sum_all + countMat - matrix(rep(sum_cat, nrow(countMat)), nrow = nrow(countMat), byrow = T)  - matrix(rep(sum_trans, ncol(countMat)), ncol = ncol(countMat))
  
  fisher_res = countMat
  for(i in 1:nrow(countMat)){
    for(j in 1:ncol(countMat)){
      fisherMat = matrix(c(
        countMat[i,j], minor_trans[i,j],
        minor_cat[i,j], minor_all[i,j]
      ), 2,2)
      fisher_res[i,j] = (fisher.test(fisherMat, alternative = alternative))$p.value
    }
  }
  return(fisher_res)
}

local_calculate_fishers = function(gene_name, expression_matrix, usage_matrix, cell_info){
# cutoffs according to p02.usage2labelcounts.R
  # 
  cell_ttlcounts_per_cluster = table(cell_info$characteristics..majorCluster)
  # get txs names
  chn_txs = rownames(usage_matrix) %>% grep(pattern = paste0("^", gene_name, "-\\d+"), value = T)
  if(length(chn_txs) > 1) {
    chn_usage_matrix = usage_matrix[chn_txs,]
    chn_expression_matrix = expression_matrix[chn_txs, ] 
  } else {
    chn_usage_matrix = usage_matrix[chn_txs, ] %>% as.matrix() %>% t()
    rownames(chn_usage_matrix) = chn_txs
    chn_expression_matrix = expression_matrix[chn_txs, ] %>% as.matrix() %>% t()
    rownames(chn_expression_matrix) = chn_txs
  }
  
  chn_expressedMat = chn_expression_matrix > transcript_express_cutoff
  chn_usage_matrix[!chn_expressedMat] = 0
  groups = cell_info$characteristics..majorCluster[match(colnames(chn_usage_matrix), cell_info$title)]
  cell_tags = apply((chn_usage_matrix > usage_cutoff), 2, function(x){
    tx = chn_txs[x]
    if(length(tx) == 0) 
      return("") 
    else 
      return(tx)
    })
  
  chn_countsMat = ((chn_usage_matrix > usage_cutoff) / 1) %>% t() %>% rowsum(., group = groups) %>% t()
  no_express_cells = apply(!chn_expressedMat, 2, all)
  assert_that(all(cell_tags[no_express_cells] == ""))
  cell_tags[no_express_cells] = paste0(gene_name, "-NO_EXPR")
  
  no_expr_row = no_express_cells %>% tapply(., INDEX = groups, sum)
  chn_countsMat = rbind(chn_countsMat, no_expr_row)
  rownames(chn_countsMat)[nrow(chn_countsMat)] = paste0(gene_name, "-NO_EXPR")
  # mix_row = apply((chn_usage_matrix <= usage_cutoff) , 2, all) %>% tapply(., INDEX = groups, sum) - no_expr_row
  mix_row = cell_ttlcounts_per_cluster - base::colSums(chn_countsMat)
  chn_countsMat = rbind(chn_countsMat, mix_row)
  rownames(chn_countsMat)[nrow(chn_countsMat)] = paste0(gene_name, "-MIX")
  
  # calculate fisher matrix
  chn_fishers = -log10(counts2FISHER_p(chn_countsMat) %>% p.adjust.matrix()) %>% round(digits = 2)
  
  # identify mix type
  mix_noexprcells = apply((chn_usage_matrix <= usage_cutoff) , 2, all)
  mix_cells = mix_noexprcells & (!no_express_cells)
  if(nrow(chn_usage_matrix) > 1 ){
    mix_usage_matrix = chn_usage_matrix[, mix_cells]
  } else{
    mix_usage_matrix = chn_usage_matrix[, mix_cells] %>% as.matrix() %>% t()
  }
  invisible(assert_that(all(chn_txs == rownames(mix_usage_matrix))))
  
  if(sum(mix_cells) > 1){
    mix_types = apply(mix_usage_matrix, 2, function(x){
    # take first two txs as mix type
      first_two = x %>% sort(decreasing = T) %>% head(2) %>% names() %>% sort() %>% paste0(collapse = "|")
      return(paste0(gene_name, "-MIX:", first_two))
    }) 
  } else if(sum(mix_cells) == 1){ 
    first_two = mix_usage_matrix %>% sort(decreasing = T) %>% head(2) %>% names() %>% sort() %>% paste0(collapse = "|")
    mix_types = paste0(gene_name, "-MIX:", first_two)
  } else { # 0
    mix_types = NULL
  }
  
  cell_tags[names(mix_types)] = mix_types
  
  return(list(count = chn_countsMat, fisher.p.adj = chn_fishers, cell_tags = cell_tags))
}

p.adjust.matrix = function(p, ...){
  stopifnot(is.matrix(p))
  p.adjs = p.adjust(p, ...)
  p.adj.matrix = matrix(
    p.adjs, 
    ncol = ncol(p),
    dimnames = dimnames(p)
    )
  return(p.adj.matrix)
}

```

### Enrichment calculation
```{r calculate all enrichment infos}
chosen_genes = exprMat %>% rownames() %>% trans2gene() %>% unique()

step = 100
cores_to_use = 10
steps = c(seq(1, length(chosen_genes), by = step), length(chosen_genes)+1)
iter01 = ".I1"
initiatePB(iter01)
all_obj_lst = list()
for(i in 2:length(steps)){
# for(i in 1:length(chosen_genes)){
  t_chosen_genes = chosen_genes[steps[i-1]: (steps[i] -1)]
  obj_lst = mclapply(t_chosen_genes, mc.cores = cores_to_use, FUN = function(chosen_gene){
  # chosen_gene = chosen_genes[i]
  obj = local_calculate_fishers(gene_name = chosen_gene, expression_matrix = exprMat, usage_matrix = usageMat, cell_info = pCells)
    return(obj)
  })
  names(obj_lst) = t_chosen_genes
  all_obj_lst = c(all_obj_lst, obj_lst)
  
  processBar(iter01, i, length(steps), tail = "ETA")
}
# names(all_obj_lst) = chosen_genes

# identify txs with ???
canonical_txs = mclapply(
  all_obj_lst, mc.cores = cores_to_use, 
  function(x){ 
    (x$count %>% apply(1, sum) %>% sort(decreasing = T) %>% names())[1] 
    }) %>% unlist()
canonical_txs = canonical_txs[grep(canonical_txs, pattern = "NO_EXPR|MIX", invert = T)]

enrichment_tbl = do.call(rbind, mclapply(all_obj_lst, mc.cores = cores_to_use, function(x){
  return(x$fisher.p.adj %>% melt(value.name = "enrich_score", varnames = c("id", "cluster")) %>% as_tibble(stringsAsFactors = F) %>% dplyr::filter(enrich_score > -log10(0.05)) %>% mutate_if(is.factor, as.character))
}))
enrichment_tbl = enrichment_tbl %>% dplyr::mutate(gene_name = trans2gene(id)) %>% dplyr::distinct() %>% 
  dplyr::filter(!grepl(id, pattern = "NO_EXPR"))
rm(list = grep(ls(), pattern = "^t_", value = T))
```

```{r plot tile, fig.height=8, fig.width=8}
# heatmap for counting
plt_tbl_enrichemnts = enrichment_tbl %>% group_by(gene_name, id, cluster) %>% dplyr::summarise(counts = n()) %>%
  group_by(gene_name) %>% summarise(id_count = length(unique(id)), 
                                    cluster_count = length(unique(cluster))) %>% 
  dplyr::mutate(id_count_cat = ifelse(id_count <= 6, as.character(id_count), ">6")) %>% 
  group_by(cluster_count, id_count_cat) %>% summarise(counts = n())
plt_tbl_enrichemnts$cluster_count = as.factor(plt_tbl_enrichemnts$cluster_count)

p_htmp = ggplot(data = plt_tbl_enrichemnts, aes(x = cluster_count, y = id_count_cat)) + 
  geom_tile(aes(fill = log10(counts))) +
  geom_text(aes(label = counts)) +
  scale_fill_gradient(low = "white", high = "#56B1F7") +
  labs(x = "# of clusters that genes are enriched in", y = "# of expression type enriched", fill = "Log10(gene count)") +
  theme(axis.line = element_blank(),
        axis.ticks = element_blank(),
        legend.position = "bottom")

p_top_clusters = plt_tbl_enrichemnts %>% group_by(cluster_count) %>% summarise(sum_count = sum(counts)) %>% 
  ggplot(aes(x = cluster_count, y = sum_count)) +
  geom_col(aes(fill = cluster_count), color = "black") +
  geom_text(aes(y = sum_count, label = sum_count), vjust = -.5, size = 4) +
  coord_cartesian(ylim = c(0, 5000))+
  labs(x = "", y = "Gene count") +
  theme(legend.position = "none", axis.text.x = element_blank())
p_right_ids = plt_tbl_enrichemnts %>% group_by(id_count_cat) %>% summarise(sum_count = sum(counts)) %>% 
  ggplot(aes(x = id_count_cat, y = sum_count)) + 
  geom_col(aes(fill = id_count_cat), color = "black") +
  geom_text(aes(y = sum_count, label = sum_count), hjust = -.4, size = 4) +
  coord_flip(ylim = c(1, 5000)) +
  labs(x = "", y = "Gene count") +
  theme(legend.position = "none", axis.text.y = element_blank(), 
        axis.text.x = element_text(angle = 45, hjust = 1))
  

require(patchwork)
p_top_clusters + plot_spacer() + p_htmp + p_right_ids + 
  plot_layout(ncol = 2, nrow = 2, heights = c(1,3), widths = c(3,1))

# pdf(file = "/home/jason/Storage/SYNC2/01.DATA/01.ASinLung/03.graphs/04.publishable/FigSupp_CD8_enrichmentCounts.pdf", height = 8, width = 8)

p_top_clusters + plot_spacer() + p_htmp + p_right_ids + 
  plot_layout(ncol = 2, nrow = 2, heights = c(1,3), widths = c(3,1))
dev.off()
# 
loginfo("# of genes with enrichment: ", length(unique(enrichment_tbl$gene_name)))
# extract genes with more than 1 tx and 1 cluster ----
genes_morethan1enr_tbl = enrichment_tbl %>% group_by(gene_name, id, cluster) %>% summarise(counts = n()) %>% 
  group_by(gene_name) %>% summarise(id_count = length(unique(id)), cluster_count = length(unique(cluster))) %>% 
  dplyr::mutate(filter_flag = id_count > 1 & cluster_count > 1)

# ggplot(genes_morethan1enr_tbl, aes(x = filter_flag)) +
#   geom_bar()+
#   geom_text(stat = "count", aes(y = ..count.., label = ..count..), vjust = -.5) +
#   labs(x = "Genes enriched more than 1 txs in more than 1 clusters")

enrichment_filter_tbl = enrichment_tbl %>% 
  dplyr::filter(gene_name %in% 
                  (genes_morethan1enr_tbl %>% dplyr::filter(filter_flag == T))$gene_name
                ) %>% distinct()
loginfo("# of genes with more than 1 enrichment in more than 1 clusters: ", length(unique(enrichment_filter_tbl$gene_name)))
```

### function & structure annotation
```{r gather annotations for enriched txs}
txs_to_annotate = (GENE.INFO %>% dplyr::filter(`Gene name` %in% enrichment_filter_tbl$gene_name))$`Transcript name` %>% unique()
# remove noncoding txs
txs_to_annotate_CD = (GENE.INFO[match(txs_to_annotate, GENE.INFO$`Transcript name`), ] %>% dplyr::filter(!is.na(`Protein stable ID`)))$`Transcript name`

# functional annotation settings ----
python_dir = "/home/jason/anaconda3/bin/python"
work_dir = "/home/jason/git/01.ASinLung/03.python/ProteinFunctionPrediction/"
getseq_dir = paste0(work_dir, "ensembl_rest_client.py")
getfunc_dir = paste0(work_dir, "iprscan5_urllib3.py")
res_path = WhereAmI("01.DATA/01.ASinLung/06.function_annotation/0809/fasta/")

# annotating ----
iter01 = ".I1"
initiatePB(iter01)
step = 100
steps = c(seq(1, length(txs_to_annotate_CD), by = step), length(txs_to_annotate_CD)+1)
for(i in 2:length(steps)){
  t_txs_toannotate = txs_to_annotate_CD[steps[i-1]:(steps[i]-1)]
  t_status = mclapply(t_txs_toannotate, mc.cores = ccores_to_use, function(chosen_tx) {
    tryCatch({
    seq_path = paste0(res_path, chosen_tx, ".fasta")
    if(file.exists(seq_path)){
      return(T)
    }
    seq_contents = ''
    x = system2(python_dir, args = c(getseq_dir, "get_sequence", chosen_tx %>% transName2transid(), "-t", "protein"), stdout = T)
    y = x[6] %>% stringr::str_split('\"',simplify = T)
    trans_seq = y[4]
    seq_contents = paste0(seq_contents,
                         '> ' , chosen_tx, " | ", chosen_tx %>% transName2transid(), "\n",
                         trans_seq, "\n")
    # write file 
    readr::write_file(seq_contents, seq_path)
    return(T)
    }, error = function(e) return(F))
    }
    )
  if(length(t_txs_toannotate[!t_status %>% unlist]) > 0){
    warning(paste0(t_txs_toannotate[!t_status %>% unlist], collapse = ","), " extraction failed!")
  }
  processBar(iter01, i, length(steps), tail = "ETA") 
}

# After running shell commands ...
  
# annother way ---- may be refined later...
# ensembl = useMart("ensembl",dataset="hsapiens_gene_ensembl")
# retrieve_sequences = biomaRt::getBM(
#   attributes = "peptide",
#   filters = "ensembl_transcript_id",
#   values = txs_to_annotate_info$`Transcript stable ID`,
#   mart = ensembl
# )
```

```{shell interpro api batch script}
to_run="txs_to_run_part01"; cat $to_run | while read file ; do /home/jason/git/01.ASinLung/03.python/ProteinFunctionPrediction/iprscan5_urllib3.py --outfile=/home/jason/Storage/SYNC2/01.DATA/01.ASinLung/06.function_annotation/0802/tsv/${file} --outformat=tsv --email=lijxug@126.com --goterms --pathways --quiet /home/jason/Storage/SYNC2/01.DATA/01.ASinLung/06.function_annotation/0802/fasta/${file} || echo $file >> ${to_run}.errlog ; done; serverChan -t "protein_annotations_complete!"
```

```{r loading realted informations}
# for transcript structure
STRUCT.INFO = readr::read_tsv(paste0(wd, "/../IDEA_data/02.IDEAinputs/GRCh38.p12.biomart.transcripts.structure.txt"))
STRUCT.INFO = STRUCT.INFO %>% dplyr::mutate(`Transcript name` = GENE.INFO$`Transcript name`[match(`Transcript stable ID`, GENE.INFO$`Transcript stable ID`)])

local_loadDomains = function(annotation_tsvs, process = T){
  domain_info_list = list()
  iter02 = ".I2"
  initiatePB(iter02)
  for (i in 1:length(annotation_tsvs)) {
    tmp_file = annotation_tsvs[i]
    trans_name = tmp_file %>% stringr::str_replace(".*//(.*)\\.fasta.*", replacement = "\\1")
    if(file.exists(tmp_file)){
      trans_info = suppressMessages(suppressWarnings(readr::read_tsv(
      tmp_file, col_names = F, progress = F
      )))
      if (nrow(trans_info)){
        trans_info = trans_info[, 1:10] # discard columnes out of 10
        colnames(trans_info)[1:10] = c(
        "Transcript name", "unknown01", "total_length", "database", "item", "description", "begin", "end", "unknown02", "unknown03"
        )
        domain_info_list[[trans_name]] = trans_info
      }
    } else { 
      domain_info_list[[trans_name]] = NULL
    }
    if(process)
      processBar(iter02, i, length(annotation_tsvs), tail = "ETA")
  }
  
  if(length(domain_info_list)){
    # combine tbl for chosen genes
    domain_info_tbl = do.call(rbind, domain_info_list) %>% as_tibble()
    # tidy df and adjust it to fit data structure for 
    domain_info_tbl = domain_info_tbl %>% dplyr::select(-matches("unknown")) %>%
    dplyr::mutate(length = end - begin, order = as.integer(as.factor(`Transcript name`)))
    return(domain_info_tbl) 
  } else{
    warning("local_loadDomains: nothing has been loaded!")
    return(tibble())
  }
}
```

```{r function/structure feature annotate}
tsv_path = paste0(wd, "/../IDEA_data/05.annotations/tsv/")
res_path = paste0(wd, "/../IDEA_data/05.annotations/fasta/")
tsv_files = Sys.glob(paste0(tsv_path, "*tsv*"))
all_domain_info_tbl = local_loadDomains(tsv_files)

# original detail tibble 
txs_detail_tbl = GENE.INFO[match(txs_to_annotate, GENE.INFO$`Transcript name`), ]

# transcript structure annotation ----
exons_tbl = STRUCT.INFO %>% group_by(`Transcript name`, `Transcript stable ID`) %>% summarise(exons = paste0(`Exon stable ID`, collapse = "|"))

txs_detail_tbl = txs_detail_tbl %>% 
  dplyr::mutate(
     `Exon IDs` = exons_tbl$exons[match(`Transcript name`, exons_tbl$`Transcript name`)], 
     Strand = STRUCT.INFO$Strand[match(`Transcript name`, STRUCT.INFO$`Transcript name`)]
  )

# protein structure annotation ----
# SignalP, Transmembrane, Pfam domain
signalP_tbl = all_domain_info_tbl %>% dplyr::select("Transcript name", "database", "item", "description") %>% distinct() %>% dplyr::filter(database == "SignalP_EUK")
transM_tbl = all_domain_info_tbl %>% dplyr::filter(database == "TMHMM") %>% dplyr::select("Transcript name", "database", "item", "begin", "end") %>% distinct() %>% group_by(`Transcript name`) %>% summarise(TMM_count = n())
pfam_tbl = all_domain_info_tbl %>% dplyr::filter(database == "Pfam") %>% dplyr::select("Transcript name", "database", "item", "begin", "end", "length") %>% distinct() %>% group_by(`Transcript name`) %>% summarise(pfam_domains = paste0(sort(item), collapse = "|"))
  
txs_detail_tbl = txs_detail_tbl %>% 
  dplyr::mutate(
    SignalP = signalP_tbl$item[match(`Transcript name`, signalP_tbl$`Transcript name`)],
    Transmembrane_number = transM_tbl$TMM_count[match(`Transcript name`, transM_tbl$`Transcript name`)],
    Pfam_domains = pfam_tbl$pfam_domains[match(`Transcript name`, pfam_tbl$`Transcript name`)]
  )

# fill 0 in coding Txs
is_coding_txs = txs_detail_tbl$`Protein stable ID` %>% is.na() %>% `!`
no_signalP_txs = is.na(txs_detail_tbl$SignalP)
no_TM_txs = is.na(txs_detail_tbl$Transmembrane_number)

txs_detail_tbl$SignalP[is_coding_txs & no_signalP_txs] = ""
txs_detail_tbl$Transmembrane_number[is_coding_txs & no_TM_txs] = 0

# add bioType
txs_detail_tbl = txs_detail_tbl %>% dplyr::mutate(bioType = type2bio(`Transcript type`))

# add protein seq
fasta_files = Sys.glob(paste0(res_path, "*.fasta"))
txs_names = stringr::str_extract(fasta_files, pattern = "(?<=/)\\w+-\\d{3}(?=\\.fasta)")
protein_seqs = mclapply(fasta_files, mc.cores = cores_to_use, function(x){
  read_tsv(x, col_types = "c") %>% as.character()
}) %>% unlist()
names(protein_seqs) = txs_names

txs_detail_tbl = txs_detail_tbl %>% dplyr::mutate(prot_seq = protein_seqs[`Transcript name`])

# add subcellular location ----
local_determineSubcellular = function(SignalP, Transmembrane_number){
  return(if_else(Transmembrane_number != 0, "membrane", 
          if_else(SignalP != "" & Transmembrane_number == 0, "secreted", 
                  if_else(SignalP == "" & Transmembrane_number == 0, "cytoplasmic", "others"))))
}
txs_detail_tbl = txs_detail_tbl %>% dplyr::mutate(
  subcellular = local_determineSubcellular(SignalP, Transmembrane_number))

```

### Enrichment changing events detection
```{r change events extraction & annotation}
local_clusterSpec_txs = function(x, t_tbl){
  sapply(x, function(x){
    t_tbl$id[t_tbl$cluster == x] %>% paste0(collapse = "|")
  })
}

local_expand = function(.tbl){
  # hard coded col_name
  col_name1 = "C1_Tx1"
  col_name2 = "C2_Tx2"
  t_repeatedTx_lst1 = .tbl[[col_name1]] %>% strsplit(split = "|", fixed = T)
  t_repeatedTx_lst2 = .tbl[[col_name2]] %>% strsplit(split = "|", fixed = T)
  
  res_tbl = tibble()
  for(k in 1:nrow(.tbl)){
    single_row = .tbl[k, ]
    t_tx1 = t_repeatedTx_lst1[[k]] %>% sort()
    t_tx2 = t_repeatedTx_lst2[[k]] %>% sort()
    common_txs = intersect(t_tx1, t_tx2)
    spec_tx1 = setdiff(t_tx1, t_tx2)
    spec_tx2 = setdiff(t_tx2, t_tx1)
    
    t_col1 = rep(spec_tx1, each = length(t_tx2))
    t_col2 = rep(t_tx2, times = length(spec_tx1))
    t_col2 = c(t_col2, rep(spec_tx2, each = length(t_tx1)))
    t_col1 = c(t_col1, rep(t_tx1, times = length(spec_tx2)))
    t_df = data.frame(t_col1, t_col2, stringsAsFactors = F) %>% distinct()
    dup_single_row = single_row[rep(1, nrow(t_df)), ]
    dup_single_row[[col_name1]] = t_df[,1]
    dup_single_row[[col_name2]] = t_df[,2]
    res_tbl = bind_rows(res_tbl, dup_single_row)
  }
  
  return(res_tbl)
}
# extract change events gene by gene
enrichment_event_tbl = tibble()
chosen_genes = enrichment_filter_tbl$gene_name %>% unique()
initiatePB(iter01)
for(i in 1:length(chosen_genes)){
  chosen_gene = chosen_genes[i]
  t_tbl = enrichment_filter_tbl %>% dplyr::filter(gene_name == chosen_gene)
  clusters = t_tbl$cluster %>% unique() %>% sort()
  
  # recombination 
  cluster_pairs = combn(clusters, 2) %>% t() 
  colnames(cluster_pairs) = c("Cluster1", "Cluster2")
  t_reformat_enrichment_tbl = as_tibble(cluster_pairs)
  t_reformat_enrichment_tbl = t_reformat_enrichment_tbl %>% dplyr::mutate(
    gene_name = t_tbl$gene_name[1],
    C1_Tx1 = local_clusterSpec_txs(Cluster1, t_tbl),
    C2_Tx2 = local_clusterSpec_txs(Cluster2, t_tbl)
  ) %>% dplyr::filter(C1_Tx1 != C2_Tx2) # remove exactly same enrichment in two cluster
  
  # expand each row to make them one tx pairs at a time
  if(nrow(t_reformat_enrichment_tbl)){
    t_reformat_enrichment_tbl = local_expand(t_reformat_enrichment_tbl)
    # sum up
    enrichment_event_tbl = bind_rows(enrichment_event_tbl, t_reformat_enrichment_tbl)
  }
  
  processBar(iter01, i, length(chosen_genes), tail = "ETA")
}
rm(cluster_pairs)

# remove same enriched items
enrichment_event_tbl = enrichment_event_tbl %>% dplyr::filter(C1_Tx1 != C2_Tx2)

# annotate enrichment changing events ----
local_sortPairs = function(x, y, sep = ":"){
  stopifnot(length(x) == length(y))
  return(lapply(1:length(x), function(i){paste0(sort(c(x[i], y[i])), collapse = ":")}) %>% unlist())
}

enrichment_event_tbl = enrichment_event_tbl %>% dplyr::mutate(`Transcript pairs` = local_sortPairs(`C1_Tx1`, `C2_Tx2`))


feature_tocompare = c("Transcript type", "bioType", "Transcription start site (TSS)", "Exon IDs", "SignalP", "Transmembrane_number", "Pfam_domains", "prot_seq", "subcellular")

Tx1_details_tbl = txs_detail_tbl[match(enrichment_event_tbl$C1_Tx1, txs_detail_tbl$`Transcript name`), ] %>% 
  dplyr::select_(.dots = paste0("`", feature_tocompare, "`"))
Tx2_details_tbl = txs_detail_tbl[match(enrichment_event_tbl$C2_Tx2, txs_detail_tbl$`Transcript name`), ] %>% 
  dplyr::select_(.dots = paste0("`", feature_tocompare, "`"))

colnames(Tx1_details_tbl) = paste0("Tx1-", colnames(Tx1_details_tbl))
colnames(Tx2_details_tbl) = paste0("Tx2-", colnames(Tx2_details_tbl))

enrichment_event_ann_tbl = bind_cols(
  enrichment_event_tbl, 
  Tx1_details_tbl, Tx2_details_tbl
)

local_detectChanges = function(df, compare_var){
  quo_varNew = paste0("Change-", compare_var)
  compare_var1 = paste0("Tx1-", compare_var)
  compare_var2 = paste0("Tx2-", compare_var)
  compare_var1 = rlang::sym(compare_var1)
  compare_var2 = rlang::sym(compare_var2)
  res_data = df %>% dplyr::mutate(
     !!quo_varNew := (!!compare_var1 != !!compare_var2)
  )
  return(res_data)
}

# common procedure to compare changes (just see if they are the same)
initiatePB(iter01)
for(i in 1:length(feature_tocompare)){
  feature = feature_tocompare[i]
  enrichment_event_ann_tbl = local_detectChanges(enrichment_event_ann_tbl, feature)
  processBar(iter01, i, length(feature_tocompare))
}
enrichment_event_ann_tbl$`Gene name` = enrichment_event_ann_tbl$gene_name

# pfam domain changes detected by annother procedure 
enrichment_event_ann_tbl = enrichment_event_ann_tbl %>% dplyr::mutate(`Change-Pfam_domains_details` = as.character(`Change-Pfam_domains`))
pairs_pfmChanged_idx = enrichment_event_ann_tbl$`Change-Pfam_domains` & !is.na(enrichment_event_ann_tbl$`Change-Pfam_domains`) 

pfmChange_tbl = enrichment_event_ann_tbl[pairs_pfmChanged_idx, ] %>% dplyr::select(`Transcript pairs`, `Tx1-Pfam_domains`, `Tx2-Pfam_domains`)

pfmChanges = pfmChange_tbl %>% as.matrix() %>% apply(1, function(x){
    tx1_pfm = x[2] %>% strsplit(split = "|", fixed = T) %>% unlist() 
    tx2_pfm = x[3] %>% strsplit(split = "|", fixed = T) %>% unlist() 
    tx1_spec_pfm = setdiff(tx1_pfm, tx2_pfm) %>% paste0(collapse = "|")
    tx2_spec_pfm = setdiff(tx2_pfm, tx1_pfm) %>% paste0(collapse = "|")
    return(paste0(
      "Tx1:", tx1_spec_pfm, ";",
      "Tx2:", tx2_spec_pfm
    ))
})

enrichment_event_ann_tbl$`Change-Pfam_domains_details`[pairs_pfmChanged_idx] = pfmChanges

# change the order of column names
col_names = colnames(enrichment_event_ann_tbl)
enrichment_event_ann_tbl = enrichment_event_ann_tbl %>% dplyr::select(
  `Gene name`, `Cluster1`, `Cluster2`, `C1_Tx1`, `C2_Tx2`, `Transcript pairs`, 
  starts_with("Change"), starts_with("Tx1-"), starts_with("Tx2-")
)
events_tbl_toStore = enrichment_event_ann_tbl

# export to table ----
require(openxlsx)
wb = EXWB.initiate()
wb = EXWB.writeSheet(wb, "CD8_byClusters_allEvents", events_tbl_toStore, overwrite = T)
# openxlsx::saveWorkbook(wb, file = paste0(WhereAmI("01.DATA/01.ASinLung/04.tables/"), "CD8_byClusters_allEvents_",format(x = Sys.time(), "%m%d"), ".xlsx"), overwrite = T)

# clear trash ----
rm(Tx1_details_tbl, Tx2_details_tbl, pfmChange_tbl, pfmChanges, events_tbl_toStore)
invisible(gc())
```

```{r additional feature}
# event types ----
enrichment_event_ann_tbl = enrichment_event_ann_tbl %>% 
  dplyr::mutate(
    involved_MIX = grepl(`Transcript pairs`, pattern = "MIX"), 
    involved_txs = grepl(`Transcript pairs`, pattern = "-\\d{3}")
  )
enrichment_event_ann_txs2txs_tbl = enrichment_event_ann_tbl %>% dplyr::filter(!involved_MIX)
# gene level features ----
# generally expressed genes across clusters
# enrichment_event_ann_tbl = enrichment_event_ann_tbl %>% 
#   dplyr::mutate(
#     generally_expressed = `Gene name` %in% generally_expressed_genes
#   )

# check: protein seq not changed ----

enrichment_event_ann_txs2txs_tbl = enrichment_event_ann_txs2txs_tbl %>% 
  dplyr::mutate(
    `Tx1-prot_seq` = protein_seqs[match(C1_Tx1, names(protein_seqs))],
    `Tx2-prot_seq` = protein_seqs[match(C2_Tx2, names(protein_seqs))],
    `Change-prot_seq` = (`Tx1-prot_seq` != `Tx2-prot_seq`)
)

```

```{r general category description, fig.height=6, fig.width=8}
library(venneuler)
library(yyplot)
require(ggforce)
require(UpSetR)

# classify by whether involved `MIX` `NO_EXPR` 
ggvenn = function(x, alpha = 0.5, text_size = 16, ...) {
  ## the venneuler depends on rJava
  ## maybe wrapping VennDiagram (output gList) as geom layer
  venneuler <- rvcheck::get_fun_from_pkg("venneuler", "venneuler") ## for easy installation, as many users have issue of rJava installation
  y <- venneuler(x)
  d <- data.frame(y$centers,
                  diameters = y$diameters,
                  labels = y$labels,
                  stringsAsFactors = FALSE)
  if(is.null(text_size)){
    ggplot(d) +
      geom_circle(aes_(x0 = ~x, y0 = ~y,
                       r = ~diameters/2, fill = ~labels),
                  alpha = alpha, ...) +
      coord_fixed()
  } else {
    ggplot(d) +
      geom_circle(aes_(x0 = ~x, y0 = ~y,
                       r = ~diameters/2, fill = ~labels),
                  alpha = alpha, ...) + 
      geom_text(aes_(x = ~x, y = ~y, label = ~labels), size = text_size) + 
      coord_fixed()
  }
}
event_class_lst = list(
  MIX_involved = (1:nrow(enrichment_event_ann_tbl))[enrichment_event_ann_tbl$involved_MIX],
  Tx_involved = (1:nrow(enrichment_event_ann_tbl))[enrichment_event_ann_tbl$involved_txs]
)

event_class_df = fromList(event_class_lst)
my_cols = gg_color_hue(2)
names(my_cols) = (event_class_df %>% colnames())[c(2,1)]
p_venn = ggvenn(event_class_df, color = NA, text_size = NULL)
p_venn + 
  scale_fill_manual(values = my_cols, breaks = c("Tx_involved", "MIX_involved")) + 
  theme_void() + theme(legend.position = "none")
upset(event_class_df, order.by = "freq", sets.bar.color = my_cols, text.scale = 1.5)

```

```{r gene level category}
# gene_level_lst = list(
#   Isoform_differentially_used_genes = enrichment_event_ann_tbl$`Gene name` %>% unique(),
#   generally_expressed_genes = generally_expressed_genes)
# upset(fromList(gene_level_lst), text.scale = 1.5)
# 
# enr_generally_genes = enrichment_event_ann_tbl$`Gene name`[enrichment_event_ann_tbl$`Gene name` %in% generally_expressed_genes] %>% unique()
# enr_generally_genes
```


```{r background txs change events}
# summation for all txs pairs ----
# all_expr_txs_tbl = tibble(id = rownames(exprMat), `Gene name` = trans2gene(rownames(exprMat)))
# all_expr_txs_tbl = all_expr_txs_tbl %>% dplyr::mutate(bioType = pGenes$bioType_tx[match(id, pGenes$transcript_name)]) %>% dplyr::filter(bioType == "protein_coding")
# gene_names = (all_expr_txs_tbl %>% group_by(`Gene name`) %>% summarise(counts = n()) %>% dplyr::filter(counts > 2))$`Gene name`
# all_expr_txs_tbl = all_expr_txs_tbl %>% dplyr::filter(`Gene name` %in% gene_names)
# gene_names = all_expr_txs_tbl$`Gene name` %>% unique()
# 
# all_coding_txs_pairs_tbl = tibble()
# initiatePB(iter01)
# for(i in 1:length(gene_names)){
#   gene_name = gene_names[i]
#   t_tbl = all_expr_txs_tbl %>% dplyr::filter(`Gene name` == gene_name)
#   t_txs = t_tbl$id
#   t_txs_pairs = combn(t_txs,2) %>% t() 
#   colnames(t_txs_pairs) = c("T1", "T2")
#   t_pairs_tbl = bind_cols(as_tibble(t_txs_pairs), `Gene name` = t_txs_pairs[,"T1"] %>% trans2gene()) %>% dplyr::arrange(T1)
#   all_coding_txs_pairs_tbl = bind_rows(all_coding_txs_pairs_tbl, t_pairs_tbl)
#   processBar(iter01, i, length(gene_names), tail = "ETA")
# }
```


```{r biotype changes, fig.height=8, fig.width=8}
enrichment_event_ann_txs2txs_tbl = enrichment_event_ann_txs2txs_tbl %>% filter(`Tx1-bioType` %in% c("protein_coding", "long_nc"), `Tx2-bioType` %in% c("protein_coding", "long_nc"))

t_chosen_genes = enrichment_event_ann_txs2txs_tbl$`Gene name` %>% unique()
t_enrichment_filter_tbl = enrichment_filter_tbl %>% filter(gene_name %in% t_chosen_genes) %>% 
  filter(!grepl(id, pattern = "MIX")) %>% 
  mutate(bioType = txs_detail_tbl$bioType[match(id, txs_detail_tbl$`Transcript name`)])
(t_enrichment_filter_tbl %>% dplyr::select(id, bioType) %>% distinct())$bioType %>% table()
pGenes$Gene.type[match(t_chosen_genes, pGenes$gene_name)] %>% table()

t_biotype_change_tbl = enrichment_event_ann_txs2txs_tbl %>%
  group_by(`Tx1-bioType`, `Tx2-bioType`) %>% summarise(count = n()) %>% 
  mutate(item = local_sortPairs(`Tx1-bioType`, `Tx2-bioType`,sep = "-")) %>% 
  group_by(item) %>% summarise(count = sum(count))

t_biotype_change_tbl %>% 
  ggplot() +
  geom_col(aes(x = item, y = count), fill = "white", color = "black") +
  geom_text(aes(x = item, y = count, label = count), vjust = -.4, size = 6) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  labs(x = "", y = "Event count")
  
# export to table ----
# wb = EXWB.writeSheet(wb, "biotypeChangingEvents",
#                      enrichment_event_ann_txs2txs_tbl %>% dplyr::filter(`Change-bioType`), overwrite = T)
```

#### Describe events by the changes
```{r potential functional changes}
require(ggrepel)
require(ggforce)

# potential functional impact of all changes between coding isforms ----
enr_coding_events_tbl = enrichment_event_ann_txs2txs_tbl %>% dplyr::filter(`Tx1-bioType` == "protein_coding" & `Tx2-bioType` == "protein_coding") 

prot_seq_change_events_tbl = enr_coding_events_tbl %>% dplyr::filter(`Change-prot_seq`)


# create a vennpie ----
local_addpositionForpie = function(x,  arc_start = 0, arc_end = 2 * pi){
  require(dplyr)
  x = x %>% dplyr::mutate(
    end = (arc_end - arc_start)*cumsum(amount)/sum(amount) + arc_start)
  starts = lag(x$end, default = arc_start)
  x = x %>% dplyr::mutate(
    start = starts
  )
  return(x)
}

prot_seq_pie_tbl = tibble(
  item = c("Same protein", "Different protein"), 
  amount = c(sum(!enr_coding_events_tbl$`Change-prot_seq`, na.rm = T), sum(enr_coding_events_tbl$`Change-prot_seq`, na.rm = T)),
  r0 = 0,
  r = 1
)
prot_seq_pie_tbl = local_addpositionForpie(prot_seq_pie_tbl)

TSS_pie_tbl = tibble(
  item = c("Same TSS", "Different TSS"),
  amount = c(sum(!(enr_coding_events_tbl %>% dplyr::filter(!`Change-prot_seq`))$`Change-Transcription start site (TSS)`), sum((enr_coding_events_tbl %>% dplyr::filter(!`Change-prot_seq`))$`Change-Transcription start site (TSS)`)), 
  r0 = 1, 
  r = 1.5
)

TSS_pie_tbl = local_addpositionForpie(
  TSS_pie_tbl, 
  arc_start = prot_seq_pie_tbl$start[with(prot_seq_pie_tbl, item == "Same protein")],
  arc_end = prot_seq_pie_tbl$end[with(prot_seq_pie_tbl, item == "Same protein")]
)
total_pie_tbl = bind_rows(prot_seq_pie_tbl, TSS_pie_tbl)


subcellular_pie_tbl = tibble(
  item = c("Subcellular location changed", "Subcellular location unchanged"),
  amount = c(sum(prot_seq_change_events_tbl$`Change-subcellular`), sum(!prot_seq_change_events_tbl$`Change-subcellular`)), 
  r0 = 1.5, 
  r = 2
)
subcellular_pie_tbl = local_addpositionForpie(
  subcellular_pie_tbl, 
  arc_start = prot_seq_pie_tbl$start[with(prot_seq_pie_tbl, item == "Different protein")],
  arc_end = prot_seq_pie_tbl$end[with(prot_seq_pie_tbl, item == "Different protein")]
)
total_pie_tbl = bind_rows(total_pie_tbl, subcellular_pie_tbl)

domain_pie_tbl = tibble(
  item = c("Lack domain infos", "Domain changed", "Domain unchanged"),
  amount = c(
    sum(is.na((prot_seq_change_events_tbl %>% dplyr::filter(`Change-subcellular`))$`Change-Pfam_domains`)), 
    sum((prot_seq_change_events_tbl %>% dplyr::filter(`Change-subcellular`))$`Change-Pfam_domains`, na.rm = T), 
    sum(!(prot_seq_change_events_tbl %>% dplyr::filter(`Change-subcellular`))$`Change-Pfam_domains`, na.rm = T)), 
  r0 = 2, 
  r = 2.5
)
domain_pie_tbl = local_addpositionForpie(
  domain_pie_tbl,
  arc_start = total_pie_tbl$start[with(total_pie_tbl, item == "Subcellular location changed")],
  arc_end = total_pie_tbl$end[with(total_pie_tbl, item == "Subcellular location changed")]
)
total_pie_tbl = bind_rows(total_pie_tbl, domain_pie_tbl)

domain_pie_tbl = tibble(
  item = c("Lack domain infos", "Domain changed", "Domain unchanged"),
  amount = c(
    sum(is.na((prot_seq_change_events_tbl %>% dplyr::filter(!`Change-subcellular`))$`Change-Pfam_domains`)), 
    sum((prot_seq_change_events_tbl %>% dplyr::filter(!`Change-subcellular`))$`Change-Pfam_domains`, na.rm = T), 
    sum(!(prot_seq_change_events_tbl %>% dplyr::filter(!`Change-subcellular`))$`Change-Pfam_domains`, na.rm = T)), 
  r0 = 2, 
  r = 2.5
)
domain_pie_tbl = local_addpositionForpie(
  domain_pie_tbl,
  arc_start = total_pie_tbl$start[with(total_pie_tbl, item == "Subcellular location unchanged")],
  arc_end = total_pie_tbl$end[with(total_pie_tbl, item == "Subcellular location unchanged")]
)
total_pie_tbl = bind_rows(total_pie_tbl, domain_pie_tbl)

# detailed legend 
total_pie_tbl$item = factor(
  total_pie_tbl$item,
  levels = c(
    "Protein:", 
    total_pie_tbl$item %>% grep(pattern = "protein", value = T),
    "TSS:",
    total_pie_tbl$item %>% grep(pattern = "TSS", value = T),
    "Subcellular:",
    total_pie_tbl$item %>% grep(pattern = "Subcellular", value = T),
    "Domain:",
    unique(total_pie_tbl$item %>% grep(pattern = "domain", ignore.case = T, value = T))
  )
)
my_cols = gg_color_hue(total_pie_tbl$item %>% unique() %>% length())

total_pie_tbl %>% dplyr::mutate(
  middle_theta = start + (end - start)/2,
  middle_raddi = r0 + (r-r0)/2,
  label_x = middle_raddi * sin(middle_theta),
  label_y = middle_raddi * cos(middle_theta), 
) %>% 
  ggplot() + coord_fixed() + theme_no_axes(base.theme = theme_bw(base_size = 18)) +
  geom_arc_bar(
    aes(x0 = 0, y0 = 0, r0 = r0, r = r,
        start = start, end = end, fill = item),
    color = "white"
  ) + 
  scale_fill_manual(
    values=c("white", my_cols[1:2], 
             "white", my_cols[3:4], 
             "white", my_cols[5:6],
             "white", my_cols[7:9]), drop=FALSE) +
  geom_text(aes(x = label_x, y = label_y, label = ifelse(amount > 5, amount, ""))) +
  theme(legend.title = element_blank()) +
  labs(title = "Classification isoform-isoform coding events")


# export to table ----
# wb = EXWB.writeSheet(wb, "SameproteinDiffTSSEvents",
#                      sheetData = enr_coding_events_tbl %>% dplyr::filter(!`Change-prot_seq` & `Change-Transcription start site (TSS)`), overwrite = T)
# wb = EXWB.writeSheet(wb, "SubcellularChangeEvents", sheetData =
#                        prot_seq_change_events_tbl %>%
#                        dplyr::filter(`Change-subcellular`),
#                      overwrite = T)
# wb = EXWB.writeSheet(wb, "DomainsChangeEvents", sheetData =
#                        prot_seq_change_events_tbl %>%
#                        dplyr::filter(`Change-Pfam_domains`),
#                      overwrite = T)
# wb = EXWB.writeSheet(wb, "Neither_sub&domains_changed", sheetData =
#                        prot_seq_change_events_tbl %>% dplyr::filter(!`Change-subcellular` & !`Change-Pfam_domains`), overwrite = T)
# openxlsx::saveWorkbook(wb, file = paste0(WhereAmI("01.DATA/01.ASinLung/04.tables/"), "CD8_byClusters_allEvents_",format(x = Sys.time(), "%m%d"), ".xlsx"), overwrite = T)
```

#### Calculate genes by event types
```{r}
genes_withDEisotypes = (enrichment_event_ann_txs2txs_tbl %>% filter(`Change-Transcription start site (TSS)`))$`Gene name` %>% unique

genes_withBiotype = (enrichment_event_ann_txs2txs_tbl %>% filter(`Change-bioType`))$`Gene name` %>% unique 

enr_coding_events_tbl = enrichment_event_ann_txs2txs_tbl %>% dplyr::filter(`Tx1-bioType` == "protein_coding" & `Tx2-bioType` == "protein_coding") 

genes_samePropDiffTSS = (enr_coding_events_tbl %>% filter(!`Change-prot_seq`&`Change-Transcription start site (TSS)`))$`Gene name` %>% unique() 

# genes_domainChange = enr_coding_events_tbl %>% filter(!(`Change-subcellular`)& `Change-Pfam_domains`) %>% with(`Gene name`) %>% unique()
genes_domainChange = enr_coding_events_tbl %>% filter(`Change-Pfam_domains`) %>% with(`Gene name`) %>% unique()
genes_subcellularChange = enr_coding_events_tbl %>% filter(`Change-subcellular`) %>% with(`Gene name`) %>% unique() 

genes_otherChanges = genes_withDEisotypes %>% setdiff(genes_withBiotype) %>% setdiff(genes_samePropDiffTSS) %>% setdiff(genes_domainChange) %>% setdiff(genes_subcellularChange)
```

##### GO for each category of DE isotypes
```{r GO for each category of DE isotypes, fig.width=12, fig.height=6}
require(enrichplot)
require(clusterProfiler)

genes_withDEisotypes2GO_lst = setNames(list(genes_withBiotype, genes_samePropDiffTSS, genes_subcellularChange, genes_domainChange), c("Coding/non-coding shifts", "Same protein with alternative TSS", "Subcellular tranlocation", "Domain cassette"))

ego_DEisotypes_lst = lapply(genes_withDEisotypes2GO_lst, function(x){
  enrichGO(x %>% XXXToEntrez(), OrgDb = "org.Hs.eg.db", ont="BP", readable=TRUE)
})

# save(
#   genes_withDEisotypes2GO_lst,
#   ego_DEisotypes_lst,
#   ego_DEisotypes_simplified_lst,
#   XXXToEntrez,
#   file = "/home/jason/Storage/tmp/eGenesDEisotypes.CD8.RDa"
# )
# load(file = "/home/jason/Storage/tmp/eGenesDEisotypes.CD8.RDa")
# require(parallel)

# ego_DEisotypes_simplified_lst = mclapply(ego_DEisotypes_lst, simplify, mc.cores = 4)

pdf(file = "/home/jason/Storage/SYNC2/01.DATA/01.ASinLung/03.graphs/04.publishable/GO.DEisotypes.CD8.pdf", width = 12, height = 6)

topN = 10
for(x in names(ego_DEisotypes_simplified_lst)){
  y = as_tibble(ego_DEisotypes_simplified_lst[[x]])
  idx = order(y$p.adjust) %>% head(topN)
  y = y[idx, ]
  y$Description = factor(y$Description, levels = y$Description[order(y$p.adjust, decreasing = T)])
  print(
    y %>% 
      ggplot(aes(x = Description, y = -log10(p.adjust))) + 
      geom_col(fill = "red", width = 0.6) + 
      labs(title = x, x = NULL, y = "-log10(adjusted p value)") + 
      coord_flip() + 
      geom_text(aes(label = round(..y.., 2)), hjust = -0.4, size = 6) + 
      scale_y_continuous(position = "right", expand = expand_scale(mult = c(0, .5))) + 
      theme_classic(base_size = 24, base_family = "Arial") +
      theme(axis.ticks.y = element_blank(), axis.line.y = element_blank())
  )
}

dev.off()
# emapplot(ego_DEisotypes_simplified_lst$`Coding/non-coding shifts`)
```

```{r detail categories under each category}

# How exactly did subcellular location changed
t_subcellular_tbl = prot_seq_change_events_tbl %>% select(matches("subcellular")) %>% arrange(`Tx1-subcellular`)
t_subcellular_tbl = t_subcellular_tbl[, -1]
t_subcellulars = t_subcellular_tbl$`Tx1-subcellular` %>% unique() %>% sort()
t_order1 = match(t_subcellular_tbl$`Tx1-subcellular`, t_subcellulars)
t_order2 = match(t_subcellular_tbl$`Tx2-subcellular`, t_subcellulars)
t_change_idx = t_order2 < t_order1
t_subcellular_location = t_subcellular_tbl[t_change_idx, ]$`Tx1-subcellular`
t_subcellular_tbl[t_change_idx, ]$`Tx1-subcellular` = t_subcellular_tbl[t_change_idx, ]$`Tx2-subcellular`
t_subcellular_tbl[t_change_idx, ]$`Tx2-subcellular` = t_subcellular_location


t_gtbl = t_subcellular_tbl %>% group_by(`Tx1-subcellular`, `Tx2-subcellular`) %>%
  summarise(count = n()) %>%
  as_tbl_graph()
t_gtbl %>% ggraph(layout = "kk") +
  geom_edge_loop(aes(label = count, color = count), angle_calc = "along",
    label_dodge = unit(2.5, 'mm')) +
  geom_edge_link(
    aes(label = count, color = count),
    angle_calc = 'along',
    label_dodge = unit(2.5, 'mm'),
  )+
  geom_node_label(aes(label = name), size = 8) +
  theme_no_axes(theme_bw(base_size = 20)) +
  coord_cartesian(xlim = c(-1,1.5))

# domain change statistic 
t_domain_tbl = prot_seq_change_events_tbl %>% filter(`Change-Pfam_domains`)
```

To relate all these events with functions, I have to link them to their potential functional changes and see if they can reveal something new in C1-C2-C3(as control) and C1-C6(as negative control) and C3-C4-C6 vs C5-C6 (to find new insights)

#### Describe change events by which clusters they expand
```{r across genes, fig.width=8, fig.height=8}
require(ggraph)
require(tidygraph)

# gene counts for all 
geneCounts_byClusters_tbl = enrichment_event_ann_txs2txs_tbl %>% 
  dplyr::mutate(col_pairs = local_sortPairs(Cluster1, Cluster2)) %>% 
  group_by(col_pairs) %>% summarise(gene_count = length(unique(`Gene name`))) %>% 
  tidyr::separate(col = col_pairs, c("Cluster1", "Cluster2"), sep = ":")

geneCounts_gtbl = geneCounts_byClusters_tbl %>% as_tbl_graph(directed = F)

p_graph = ggraph(geneCounts_gtbl, layout = "kk") +
  geom_edge_link(aes(label = gene_count, color = gene_count), angle_calc = "along", label_dodge = unit(2.5, 'mm'), ) +
  ggraph::scale_edge_color_gradient(name = "Gene count", low = "white", high = "red") +
  geom_node_label(aes(label = name)) + 
  theme_void() +
  coord_fixed(xlim = c(-3,3)) +
  theme(legend.position = "bottom") +
  theme(legend.position = "bottom") +
  labs(title = "")

p_graph
# ggsave(plot = p_graph, device = "pdf", filename = "/home/jason/Storage/SYNC2/01.DATA/01.ASinLung/03.graphs/04.publishable/Fig3_C6_DEisotypes.pdf", width = 8, height = 8)
# rm(geneCounts_byClusters_tbl, geneCounts_gtbl)

```

<!-- another attempt -->
<!-- ```{r} -->
<!-- library(igraph) -->
<!-- library(tcltk) -->

<!-- # net.filter <- graph_from_data_frame(d = geneCounts_byClusters_tbl, directed = F) -->
<!-- # lay <- layout_with_fr(net.filter) -->
<!-- # lay <- norm_coords(lay, ymin = -1, ymax = 1, xmin = -1, xmax = 1) -->
<!-- net <- graph_from_data_frame(d = geneCounts_byClusters_tbl, directed = F) -->
<!-- tkplot(net,  -->
<!--        edge.arrow.size = .4,  -->
<!--        vertex.label = V(net)$id,  -->
<!--        edge.curved = .2, -->
<!--        vertex.label.font = 1, -->
<!--        rescale = F,  -->
<!--        layout = lay, -->
<!--        vertex.label.cex = 1, -->
<!--        vertex.frame.color = "white") -->
<!-- ``` -->

#### simple counts for each cluster
```{r}
chosen_clusters = union(enrichment_event_ann_txs2txs_tbl$Cluster1 %>% unique(), enrichment_event_ann_txs2txs_tbl$Cluster2 %>% unique()) %>% sort()

p_isoTypes = lapply(chosen_clusters, function(x){
  with(enrichment_event_ann_txs2txs_tbl, `Gene name`[Cluster1 == x | Cluster2 == x] %>% unique() %>% length())
}) %>% unlist() %>% setNames(nm = chosen_clusters) %>% enframe() %>% 
  ggplot(aes(x = name, y = value)) +
  geom_col(width = 0.8, fill = "white", color = "black") +
  geom_text(aes(label = ..y..), vjust = -0.4, size = 5) +
  labs(x = NULL, y = "nGenes with differential isotypes") +
  scale_y_continuous(expand = expand_scale(mult = c(0, 0.2))) +
  theme(axis.text.x = element_text(angle = 30, hjust = 1))

p_isoTypes
# ggsave(plot = p_isoTypes, filename = "/home/jason/Storage/SYNC2/01.DATA/01.ASinLung/03.graphs/04.publishable/Fig3_isotypeCounts.pdf", device = pdf, )

```

After seeing this whole picture, let's focus on real biological question ...

## how exactly this change events would reveal the functional differences between clusters
### cluster comparison
```{r cluster pairs comparison}
chosen_clusters = c( "CD8_C6-LAYN")
# event_changes_tbl = enrichment_event_ann_txs2txs_tbl %>%
#   dplyr::filter(Cluster1 %in% chosen_clusters , Cluster2 %in% chosen_clusters) %>%
#   dplyr::mutate(cluster_pairs = local_sortPairs(Cluster1, Cluster2)) %>%
#   dplyr::filter(cluster_pairs %in% c(paste(chosen_clusters,collapse = ":"))) %>%
#   dplyr::select(`Gene name`, Cluster1, Cluster2, C1_Tx1, C2_Tx2, `cluster_pairs`, contains("subcellular"))

event_changes_tbl = enrichment_event_ann_txs2txs_tbl %>%
  dplyr::filter(Cluster1 %in% chosen_clusters | Cluster2 %in% chosen_clusters) %>%
  dplyr::mutate(cluster_pairs = local_sortPairs(Cluster1, Cluster2)) %>%
  dplyr::filter(grepl(cluster_pairs, pattern = chosen_clusters)) %>%
  dplyr::select(`Gene name`, Cluster1, Cluster2, C1_Tx1, C2_Tx2, `cluster_pairs`, contains("subcellular"))
# sort
event_changes_tbl = event_changes_tbl %>% 
  tidyr::separate(cluster_pairs, into = c("tCluster1","tCluster2"), sep = ":", remove = F)
# t_idx = with(event_changes_tbl, Cluster1 != tCluster1)
# event_changes_reform_tbl = within(
#   event_changes_tbl,
#   {
#     t_val = C2_Tx2[t_idx]
#     C2_Tx2[t_idx] = C1_Tx1[t_idx]
#     C1_Tx1[t_idx] = t_val
#     t_val = `Tx2-subcellular`[t_idx]
#     `Tx2-subcellular`[t_idx] = `Tx1-subcellular`[t_idx]
#     `Tx1-subcellular`[t_idx] = t_val
#     
#     t_val = Cluster2[t_idx]
#     Cluster2[t_idx] = Cluster1[t_idx]
#     Cluster1[t_idx] = t_val
#     
#     rm(t_val, tCluster1, tCluster2)
#   })

plt_alluvium_tbl = event_changes_reform_tbl %>% 
  group_by_(.dots = paste0("`",names(event_changes_reform_tbl)[matches("subcellular", vars = names(event_changes_reform_tbl))], "`")) %>% 
  summarise(counts = n())

# chosen_enrichments_tbl = 
#   enrichment_filter_tbl %>% dplyr::filter(cluster %in% chosen_clusters, id %in% chosen_txs)
# chosen_enrichments_tbl = bind_cols(
#   chosen_enrichments_tbl,
#   txs_detail_tbl[match(chosen_enrichments_tbl$id, txs_detail_tbl$`Transcript name`), "subcellular"]) %>% 
#   dplyr::mutate(isMIX = grepl(id, pattern = "MIX"))
# chosen_enrichments_tbl = bind_cols(
#   chosen_enrichments_tbl,
#   txs_detail_tbl[match(chosen_enrichments_tbl$id, txs_detail_tbl$`Transcript name`), c(
#     "Gene type", "Transcript type", "Transcription start site (TSS)", "APPRIS annotation", "SignalP","Transmembrane_number","Pfam_domains","bioType","prot_seq"
#   )]) %>% dplyr::mutate(
#     isMIX = grepl(id, pattern = "MIX")
#   )

# chosen_enrichments_counts_tbl = chosen_enrichments_tbl %>% 
#   group_by(cluster, isMIX) %>% summarise(freq = n()) %>% ungroup()
# chosen_enrichments_counts_tbl = chosen_enrichments_counts_tbl %>% dplyr::mutate(cohort = rownames(chosen_enrichments_counts_tbl))

ggplot(plt_alluvium_tbl,
       aes(axis1 = `Tx1-subcellular`, axis2 = `Tx2-subcellular`,
           y = counts)) +
  geom_alluvium(aes(fill = `Change-subcellular`)) +
  geom_stratum() +
  geom_label(stat = "stratum", label.strata = TRUE) +
  scale_x_discrete(limits = c("others", chosen_clusters), expand = c(.1, .05)) +
  theme(legend.position = "bottom")
  

```


--- inspector for each chosen genes

# Check single gene
```{r inspect single gene, fig.height=14, fig.width=10}
theme_set(theme_classic(base_size = 20))
chosen_gene = "CD8A"
total = f_plotGeneInspect(chosen_gene, annotation_path = tsv_path)
total

```

### chosen genes batch outputs
```{r batch chosen gene inspector, include=FALSE, fig.height=12, fig.width=10}
chosen_genes = c(
  "PTPRC",
  "CD8A",
  "HAPLN3",
  "PRDM1",
  "CCR6",
  "BTLA"
)

err_list = c()
plt_list = list()
for(i in 1:length(chosen_genes)){
  tryCatch({
  chosen_gene = chosen_genes[i]
  total = f_plotGeneInspect(chosen_gene, annotation_path = tsv_path)
  if(any(is.na(total))) {
    next
  }
  # ggsave(filename = paste0(i, ".", chosen_gene, ".png"), plot = total, device = "png", path = out_dir, width = 10, height = 14)
  plt_list[[chosen_gene]] = total
  loginfo(chosen_gene, " done!")
  }, error = function(e) {err_list = c(err_list, chosen_gene)})
}
print(err_list)

print(plt_list)

out_dir = paste0(
    WhereAmI("01.DATA/01.ASinLung/03.graphs/04.publishable/"), "publishGenes_",
    format(x = Sys.time(), "%m%d"), ".pdf")

pdf(file = out_dir, title = chosen_gene, width = 10, height = 14)

# msgASAP(ttl = "Graphs_output_Done")

dev.off()

```


```{r GO plots, fig.width=10, fig.height=8}
library(enrichplot)
library(clusterProfiler)
 
# check CD8 enrichment changing genes GO BP enrichment
ego = enrichGO(enrichment_event_ann_tbl$`Gene name` %>% unique() %>% XXXToEntrez(), OrgDb = "org.Hs.eg.db", ont="BP", readable=TRUE)
# goplot(ego)
# remove redundant GO terms
ego2 = clusterProfiler::simplify(ego)

# plot ----
barplot(ego2, showCategory=20)
cnetplot(ego2, showCategory = 5)
emapplot(ego2)
```

## isoform usage distribution
```{r, fig.width=8, fig.height=6}
theme_set(theme_classic(base_size = 18, base_family = ""))
require(ggridges)
t_idx_id = sample(1:nrow(usageMat), round(nrow(usageMat) * 0.1, 0))
t_cell = sample(1:ncol(usageMat), round(ncol(usageMat) * 0.1, 0))
usage_tbl = usageMat[t_idx_id, t_cell] %>% melt(varnames = c("tx_id", "cell_id"), value.name = "usage") %>% as_tibble() %>% dplyr::mutate(cluster = pCells$characteristics..majorCluster[match(cell_id, pCells$title)])
p_usage = usage_tbl %>% dplyr::filter(usage > 0) %>% 
  ggplot()+
  # geom_density_ridges2(aes(x = usage, y = cluster, fill = cluster))
  geom_histogram(aes(x = usage, fill = cluster), bins = 50) +
  facet_wrap(~cluster) +
  scale_y_log10(
    breaks = scales::trans_breaks("log10", function(x) 10^x),
    labels = scales::trans_format("log10", scales::math_format(10^.x))
  ) +
  theme_classic(base_size = 18) +
  theme(legend.position = "none") 
p_usage

# pdf(file = "/home/jason/Storage/SYNC2/01.DATA/01.ASinLung/03.graphs/04.publishable/FigSupp_isoformUsageDistribution.pdf", width = 8, height = 6)
# p_usage
# dev.off()

expr_tbl = exprMat[1, ] %>%
  as.matrix() %>% t() %>%
  melt(varnames = c("tx_id", "cell_id"), value.name = "expr") %>% as_tibble() %>% dplyr::mutate(cluster = pCells$characteristics..majorCluster[match(cell_id, pCells$title)])
p_geneExpr = expr_tbl %>% ggplot()+
  geom_density_ridges2(aes(x = log10(expr+1), y = cluster, fill = cluster))

expr_tbl = exprMat[t_idx_id, t_cell] %>%
  melt(varnames = c("tx_id", "cell_id"), value.name = "expr") %>% as_tibble() %>% dplyr::mutate(cluster = pCells$characteristics..majorCluster[match(cell_id, pCells$title)])
p_isoExpr = expr_tbl %>% dplyr::filter(expr>0) %>% ggplot()+
  geom_density_ridges2(aes(x = log10(expr+1), y = cluster, fill = cluster))

p_geneExpr+ theme(legend.position = "none")
p_isoExpr+ theme(legend.position = "none")

rm(t_idx_id, t_cell, usage_tbl, expr_tbl)
invisible(gc())
```

## Separate tmp plots
```{r, fig.height=4, fig.width=12}
# specific plot for transcript splicing diagram
chn_txs = c("RUNX3-201", "RUNX3-203", "RUNX3-205")
test_struct_tbl = STRUCT.INFO %>% dplyr::filter(`Transcript name` %in% chn_txs) %>% dplyr::mutate(bioType = type2bio(GENE.INFO$`Transcript type`[match(`Transcript name`, GENE.INFO$`Transcript name`)]))
p6 = local_drawTranscriptStrucutre(test_struct_tbl)
p6 = p6 + theme(legend.position = "right", axis.text.x = element_text(angle = 45, hjust = 1))
p6

annotation_tsvs = paste0("/data1/jason/data/01.ASinLung/tmp/tsv/", chn_txs, ".fasta.tsv.txt")
t_domain_info_tbl = local_loadDomains(annotation_tsvs)
t_domain_info_tbl = t_domain_info_tbl %>% dplyr::mutate(entryName = `Transcript name`)
p5 = local_drawProteinStructure(t_domain_info_tbl) 
p5 = p5 + theme(legend.position = "right")
p5
```

## calc exhaustion rate
```{r}
CD8cells_exhaustionGenes = read_tsv("/home/jason/Storage/SYNC2/01.DATA/01.ASinLung/00.tmp/CD8exhaustionDEGs.txt", col_names = F)$X1
# CD4cells_exhaustionGenes = read_tsv("/home/jason/Storage/SYNC2/01.DATA/01.ASinLung/00.tmp/CD4exhaustionDEGs.txt", col_names = F)$X1

chosen_idx = rownames(exprMat_sum) %in% CD8cells_exhaustionGenes
t_exhaustion_scores = apply(exprMat_sum[chosen_idx, ], 2, function(x){
  mean(log2(x+1))
})

t_cellgroups = pCells$characteristics..majorCluster[match(names(t_exhaustion_scores), pCells$title)]
names(t_cellgroups) = names(t_exhaustion_scores)

t_exhaustion_scores_byCluster = tapply(t_exhaustion_scores, INDEX = t_cellgroups, mean)

```

## Check mean usage changes
```{r, fig.height=12, fig.width=8}

chosen_gene = "PTPRC"

geneInspect_data = f_plotGeneInspect(chosen_gene, annotation_path = tsv_path, return_data = T)

t_usageMat = usageMat[grep(rownames(usageMat), pattern = paste0("^", chosen_gene, "-\\d{3}")),]
t_usageMat_tbl = t_usageMat %>% melt(varnames = c("id", "cell_id"), value.name = "usage") %>% as_tibble() %>% 
  dplyr::mutate(cluster = pCells$characteristics..majorCluster[match(cell_id, pCells$title)]) %>% 
  mutate(exhaustion_score = t_exhaustion_scores[cell_id], 
         mean_exhaustion_score = as.numeric(t_exhaustion_scores_byCluster[cluster]))

t_usageMat_tbl$id = as.character(t_usageMat_tbl$id)
# MIX_cells = all_obj_lst[["HAPLN3"]]$cell_tags %>% grep(pattern = "MIX", value = T) %>% names()
chn_clusters = t_usageMat_tbl$cluster %>% unique()
chn_id = geneInspect_data$count %>% rownames() %>% grep(invert = T, pattern = "MIX", value = T) %>% sort()
# chn_id = c("HAPLN3-201", "HAPLN3-202")

p_meanUsage = t_usageMat_tbl %>% 
  filter(id %in% chn_id) %>%
  # mutate(id = ifelse(id == "HAPLN3-201", "HAPLN3-201", "Other HAPLN3 isoforms")) %>% 
  dplyr::group_by(id, cluster, mean_exhaustion_score) %>% 
  filter(cluster %in% chn_clusters) %>%
  summarise(mean = mean(usage), n = n(), sd = sd(usage), se = sd/sqrt(n)) %>% 
  ungroup() %>% group_by(id) %>% 
  # ggplot(aes(x = cluster, y = mean, color = mean_exhaustion_score))+
  ggplot(aes(x = cluster, y = mean, color = id, fill = id))+
  # geom_bar(stat= "identity", position=position_dodge(), fill = "white") +
  geom_bar(stat= "identity", position=position_dodge()) +
  # geom_point() +
  geom_errorbar(aes(ymin=mean-se, ymax=mean+se), width=.2) +
  # geom_errorbar(aes(ymin=mean-se, ymax=mean+se), width=.2, position=position_dodge(.9)) +
  # geom_path(aes(group = id))+
  theme_classic(base_size = 24) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  facet_wrap(~id, ncol = 1) +
  # labs(y = "Mean usage level", x = "", color = "Exhaustion score")
  labs(y = "Mean usage level", x = "", color = "Isoform", fill = "Isoform")
p_meanUsage

p_meanUsage = t_usageMat_tbl %>% 
  filter(id %in% chn_id) %>%
  # mutate(id = ifelse(id == "HAPLN3-201", "HAPLN3-201", "Other HAPLN3 isoforms")) %>% 
  dplyr::group_by(id, cluster, mean_exhaustion_score) %>% 
  filter(cluster %in% chn_clusters) %>%
  summarise(mean = mean(usage), n = n(), sd = sd(usage), se = sd/sqrt(n)) %>% 
  ungroup() %>% group_by(id) %>% 
  ggplot(aes(x = cluster, y = mean, color = mean_exhaustion_score))+
  geom_bar(stat= "identity", position=position_dodge(), fill = "white") +
  geom_point() +
  geom_errorbar(aes(ymin=mean-se, ymax=mean+se), width=.2) +
  # geom_errorbar(aes(ymin=mean-se, ymax=mean+se), width=.2, position=position_dodge(.9)) +
  # geom_path(aes(group = id))+
  theme_classic(base_size = 24) +
  theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  facet_wrap(~id, ncol = 1) +
  labs(y = "Mean usage level", x = "", color = "Exhaustion score")
p_meanUsage
# pdf(file = 
#       paste0(
#         WhereAmI("01.DATA/01.ASinLung/03.graphs/04.publishable/"), chosen_gene, "_meanUsageCD8.pdf"
#       ), title = chosen_gene, width = 8, height = 6)
# p_meanUsage
# dev.off()
# t_testres = wilcox.test(
#   t_q1, t_q2
# )
```


### exhaustion rate trend

#### plot usage vs. exhaustion rate
```{r, fig.height=6, fig.width=8}
chosen_gene = "CD8A"

t_usageMat = usageMat[grep(rownames(usageMat), pattern = paste0("^", chosen_gene, "-\\d{3}")),]

t_exhaustion_tbl = t_usageMat %>% 
  melt(varnames = c("id", "cell_id"), value.name = "usage") %>% as_tibble() %>% 
  dplyr::mutate(cluster = pCells$characteristics..majorCluster[match(cell_id, pCells$title)]) %>% 
  mutate(exhaustion_score = t_exhaustion_scores[cell_id], mean_exhaustion_score = t_exhaustion_scores_byCluster[cluster])

# t_exhaustion_tbl %>% ggplot(
#   aes(x = cluster, y = exhaustion_score)
# ) +
#   geom_boxplot() +
#   theme(axis.text.x = element_text(angle = 45, hjust = 1))
# 
# t_exhaustion_tbl %>% ggplot(
#   aes(x = cluster, y = usage, fill = exhaustion_score)
# ) +
#   geom_boxplot() +
#   facet_wrap(~id, ncol = 1) +
#   theme(axis.text.x = element_text(angle = 45, hjust = 1))

t_exhaustion_tbl %>%
  filter(id == "CD8A-202") %>% group_by(cluster) %>% 
  summarise(exhaustion_score_mean = mean(exhaustion_score), usage_mean = mean(usage)) %>% 
ggplot(aes(x = exhaustion_score_mean, y = usage_mean)) +
  geom_point() +
  geom_smooth(method = "lm")

t_exhaustion_tbl %>% 
    ggplot(aes(x = cluster, y = usage, color = mean_exhaustion_score))+
  geom_boxplot() +
    geom_point(stat = "summary", fun.y = "mean") +
    geom_path(stat = "summary", fun.y = "mean", aes(group = id))+
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
  facet_wrap(~id, ncol =1) +
    labs(y = "Usage", x = "", fill = "Exhaustion score")

cor_res = cor.test(with(t_exhaustion_tbl, exhaustion_score[id == "CD8A-202"]), with(t_exhaustion_tbl, usage[id == "CD8A-202"]))

t_exhaustion_tbl %>% 
  filter(id == "CD8A-202") %>% 
  ggplot(aes(x = usage, y = exhaustion_score)) +
  geom_point() +
  geom_smooth(method = "lm") +
  labs(x = "CD8A-202 usage")

```

### MIXture determination
```{r}
theme_set(theme_classic(base_size = 18))

plot_MIXtureDistribution = function(chosen_gene, chosen_cluster, return_data = F){
  t_usageMat = usageMat[grep(rownames(usageMat), pattern = paste0("^", chosen_gene, "-\\d{3}")),]
  t_usageMat_tbl = t_usageMat %>% melt(varnames = c("id", "cell_id"), value.name = "usage") %>% as_tibble() %>% 
    dplyr::mutate(
      cluster = pCells$characteristics..majorCluster[match(cell_id, pCells$title)],
      tissue = pCells$characteristics..tissueType[match(cell_id, pCells$title)])
   
  # geneInspect_data = f_plotGeneInspect(chosen_gene, annotation_path = tsv_path, return_data = T)
  # chn_id = geneInspect_data$count %>% rownames() %>% grep(invert = T, pattern = "MIX", value = T) %>% sort()
  
  t_usageMat_tbl$id = as.character(t_usageMat_tbl$id)
  
  t_usageMat_tbl = t_usageMat_tbl %>% dplyr::mutate(
    cell_tag = all_obj_lst[[chosen_gene]]$cell_tags[cell_id])
  
  
  t_usageMat_tbl = t_usageMat_tbl %>% dplyr::filter(grepl(cluster, pattern = chosen_cluster)) %>% 
    dplyr::filter(grepl(cell_tag, pattern = "MIX")) 
  
  if(return_data){
    return(t_usageMat_tbl)
  }
  
  p_usageMix = t_usageMat_tbl %>% 
    # dplyr::filter(id %in% chn_id) %>% 
    ggplot(aes(x = id, y = usage, fill = id)) +
    geom_boxplot() +
    theme(axis.text.x = element_text(angle = 45, hjust = 1)) +
    labs(title = paste0(chosen_gene, " MIX distribution in CD8 ", chosen_cluster), y = "Usage", x = "", fill = "Isoform")
  return(p_usageMix)
}

chosen_gene = "PRDM1"
chosen_cluster = "C3"

p_usageMix = plot_MIXtureDistribution(chosen_gene, chosen_cluster)
p_usageMix

pdf(file =
      paste0(
        WhereAmI("01.DATA/01.ASinLung/03.graphs/04.publishable/"), chosen_gene, "_CD8_", chosen_cluster, "_mixture.pdf"
      ), title = chosen_gene, width = 8, height = 6)
p_usageMix

dev.off()
```

# Enrichment changing events caculation by Tissue
```{r calculation by Tissue functions}
local_calculate_fishersByTissue = function(gene_name, expression_matrix, usage_matrix, cell_info){
# cutoffs according to p02.usage2labelcounts.R
  # 
  cell_ttlcounts_per_cluster = table(cell_info$characteristics..tissueType)
  # get txs names
  chn_txs = rownames(usage_matrix) %>% grep(pattern = paste0("^", gene_name, "-\\d+"), value = T)
  if(length(chn_txs) > 1) {
    chn_usage_matrix = usage_matrix[chn_txs,]
    chn_expression_matrix = expression_matrix[chn_txs, ] 
  } else {
    chn_usage_matrix = usage_matrix[chn_txs, ] %>% as.matrix() %>% t()
    rownames(chn_usage_matrix) = chn_txs
    chn_expression_matrix = expression_matrix[chn_txs, ] %>% as.matrix() %>% t()
    rownames(chn_expression_matrix) = chn_txs
  }
  
  chn_expressedMat = chn_expression_matrix > transcript_express_cutoff
  chn_usage_matrix[!chn_expressedMat] = 0
  groups = cell_info$characteristics..tissueType[match(colnames(chn_usage_matrix), cell_info$title)]
  cell_tags = apply((chn_usage_matrix > usage_cutoff), 2, function(x){
    tx = chn_txs[x]
    if(length(tx) == 0) 
      return("") 
    else 
      return(tx)
    })
  
  chn_countsMat = ((chn_usage_matrix > usage_cutoff) / 1) %>% t() %>% rowsum(., group = groups) %>% t()
  no_express_cells = apply(!chn_expressedMat, 2, all)
  assert_that(all(cell_tags[no_express_cells] == ""))
  cell_tags[no_express_cells] = paste0(gene_name, "-NO_EXPR")
  
  no_expr_row = no_express_cells %>% tapply(., INDEX = groups, sum)
  chn_countsMat = rbind(chn_countsMat, no_expr_row)
  rownames(chn_countsMat)[nrow(chn_countsMat)] = paste0(gene_name, "-NO_EXPR")
  # mix_row = apply((chn_usage_matrix <= usage_cutoff) , 2, all) %>% tapply(., INDEX = groups, sum) - no_expr_row
  mix_row = cell_ttlcounts_per_cluster - base::colSums(chn_countsMat)
  chn_countsMat = rbind(chn_countsMat, mix_row)
  rownames(chn_countsMat)[nrow(chn_countsMat)] = paste0(gene_name, "-MIX")
  
  # calculate fisher matrix ----
  chn_fishers = -log10(counts2FISHER_p(chn_countsMat) %>% p.adjust.matrix()) %>% round(digits = 2)
  
  # identify mix type
  mix_noexprcells = apply((chn_usage_matrix <= usage_cutoff) , 2, all)
  mix_cells = mix_noexprcells & (!no_express_cells)
  if(nrow(chn_usage_matrix) > 1 ){
    mix_usage_matrix = chn_usage_matrix[, mix_cells]
  } else{
    mix_usage_matrix = chn_usage_matrix[, mix_cells] %>% as.matrix() %>% t()
  }
  invisible(assert_that(all(chn_txs == rownames(mix_usage_matrix))))
  
  if(sum(mix_cells) > 1){
    mix_types = apply(mix_usage_matrix, 2, function(x){
    # take first two txs as mix type
      first_two = x %>% sort(decreasing = T) %>% head(2) %>% names() %>% sort() %>% paste0(collapse = "|")
      return(paste0(gene_name, "-MIX:", first_two))
    }) 
  } else if(sum(mix_cells) == 1){ 
    first_two = mix_usage_matrix %>% sort(decreasing = T) %>% head(2) %>% names() %>% sort() %>% paste0(collapse = "|")
    mix_types = paste0(gene_name, "-MIX:", first_two)
  } else { # 0
    mix_types = NULL
  }
  
  cell_tags[names(mix_types)] = mix_types
  
  return(list(count = chn_countsMat, fisher.p.adj = chn_fishers, cell_tags = cell_tags))
}
```

```{r}
chosen_genes = exprMat %>% rownames() %>% trans2gene() %>% unique()

step = 1000
steps = c(seq(1, length(chosen_genes), by = step), length(chosen_genes)+1)
iter01 = ".I1"
initiatePB(iter01)
all_obj_lst_byTissue = list()
for(i in 2:length(steps)){
# for(i in 1:length(chosen_genes)){
  t_chosen_genes = chosen_genes[steps[i-1]: (steps[i] -1)]
  obj_lst = mclapply(t_chosen_genes, mc.cores = cores_to_use, FUN = function(chosen_gene){
  # chosen_gene = chosen_genes[i]
    obj = local_calculate_fishersByTissue(gene_name = chosen_gene, expression_matrix = exprMat, usage_matrix = usageMat, cell_info = pCells)
    return(obj)
  })
  names(obj_lst) = t_chosen_genes
  all_obj_lst_byTissue = c(all_obj_lst_byTissue, obj_lst)
  
  processBar(iter01, i, length(steps), tail = "ETA")
}
# names(all_obj_lst_byTissue) = chosen_genes

enrichment_tbl_byTissue = do.call(rbind, mclapply(all_obj_lst_byTissue, mc.cores = cores_to_use, function(x){
  return(x$fisher.p.adj %>% melt(value.name = "enrich_score", varnames = c("id", "cluster")) %>% as_tibble(stringsAsFactors = F) %>% dplyr::filter(enrich_score > -log10(0.05)) %>% mutate_if(is.factor, as.character))
}))
enrichment_tbl_byTissue = enrichment_tbl_byTissue %>% dplyr::mutate(gene_name = trans2gene(id)) %>% dplyr::distinct() %>% 
  dplyr::filter(!grepl(id, pattern = "NO_EXPR"))

# save only enriched more than one isoform genes in more than one tissue 
loginfo("# of genes with enrichment: ", length(unique(enrichment_tbl_byTissue$gene_name)))
# extract genes with more than 1 tx and 1 cluster ----
genes_morethan1enr_tbl_byTissue = enrichment_tbl_byTissue %>% group_by(gene_name, id, cluster) %>% summarise(counts = n()) %>% 
  group_by(gene_name) %>% summarise(id_count = length(unique(id)), cluster_count = length(unique(cluster))) %>% 
  dplyr::mutate(filter_flag = id_count > 1 & cluster_count > 1)

# ggplot(genes_morethan1enr_tbl_byTissue, aes(x = filter_flag)) +
#   geom_bar()+
#   geom_text(stat = "count", aes(y = ..count.., label = ..count..), vjust = -.5) +
#   labs(x = "Genes enriched more than 1 txs in more than 1 clusters")

enrichment_filter_tbl_byTissue = enrichment_tbl_byTissue %>% 
  dplyr::filter(gene_name %in% 
                  (genes_morethan1enr_tbl_byTissue %>% dplyr::filter(filter_flag == T))$gene_name
                ) %>% distinct()
loginfo("# of genes with more than 1 enrichment in more than 1 clusters: ", length(unique(enrichment_filter_tbl_byTissue$gene_name)))


rm(list = grep(ls(), pattern = "^t_", value = T))
# save(list = c("all_obj_lst_byTissue", "enrichment_filter_tbl_byTissue"),
#      file = "/data1/jason/data/01.ASinLung/results/EnrichmentResultsByTissue.RDa")
```
to perform cross exmaination with clone-based analysis ...

# Output dataBrowser data set 
```{r}
save(file = "/data1/jason/data/01.ASinLung/tmp/plt_data_181126.rda", list = c("pCells", "exprMat_sum_tbl", "all_obj_lst", "STRUCT.INFO", "all_obj_lst_byTissue"))
```



# exploring (xiajibashi)
```{r}
chn_txs = grep(rownames(exprMat), pattern = paste0("^", chosen_gene, "-\\d{3}"), value = T)
chn_clusters = c("CD8_C6-LAYN")
chn_cells = (pCells %>% filter(characteristics..majorCluster %in% chn_clusters))$title
exprMat[chn_txs, chn_cells] %>% melt(varnames = c("id", "cell_id"), value.name = "tpm") %>% mutate(cluster = pCells$characteristics..majorCluster[match(cell_id, pCells$title)]) %>% ggplot(aes(x = id, y = log10(tpm+1), fill = cluster)) + geom_boxplot() + theme(axis.text.x = element_text(angle = 45, hjust = 1))

# output pCell infos
t_CD8pCell_table = pCells %>% dplyr::select(characteristics..patient, characteristics..sampleType, Number.of.input.reads, Average.input.read.length, Uniquely.mapped.reads.number)
colnames(t_CD8pCell_table) = c("Patient", "SampleType", "Number of input reads", "Average input read length", "Uniquely mapped reads number")

t_CD8pCell_table = t_CD8pCell_table %>% group_by(Patient) %>% summarise(`Number of cells` = n(), `Number of input reads (Mean)` = mean(`Number of input reads`), `Average input read length (Mean)` = mean(`Average input read length`), `Uniquely mapped reads number (Mean)` = mean(`Uniquely mapped reads number`))

t_CD8pCell_wb = openxlsx::write.xlsx(t_CD8pCell_table, file = "/home/jason/Storage/SYNC2/01.DATA/01.ASinLung/00.tmp/CD8pCells.xlsx")
```

# Summary of enrichments for each cluster
```{r}
require(clusterProfiler)
require(enrichplot)

chosen_genes = enrichment_filter_tbl$gene_name %>% unique()
t_ego_enri = enrichGO(chosen_genes %>% XXXToEntrez(), OrgDb = "org.Hs.eg.db", ont="BP", readable=TRUE)
t_ego_enri2 = simplify(t_ego_enri)
dotplot(t_ego_enri2, showCategory = 10)
```
# Details about each category of events
```{r, fig.height=6, fig.width=9}
# theme_set(theme_classic(base_size = 20))
t_enrich_tbl = enrichment_event_ann_tbl %>% filter(!involved_MIX, `Change-prot_seq`, `Change-Pfam_domains`)
# gene counts for all 
geneCounts_byClusters_tbl = t_enrich_tbl %>% 
  dplyr::mutate(col_pairs = local_sortPairs(Cluster1, Cluster2)) %>% 
  group_by(col_pairs) %>% summarise(gene_count = length(unique(`Gene name`))) %>% 
  tidyr::separate(col = col_pairs, c("Cluster1", "Cluster2"), sep = ":")

geneCounts_gtbl = geneCounts_byClusters_tbl %>% as_tbl_graph(directed = F)

# set.seed(100)
ggraph(geneCounts_gtbl, layout = "kk") +
  geom_edge_link(aes(label = gene_count, color = gene_count), angle_calc = "along", label_dodge = unit(2.5, 'mm'), label_size = 6) +
  ggraph::scale_edge_color_gradient(low = "#1E90FF", high = "#FFA500") +
  geom_node_label(aes(label = name), size = 6) + 
  theme_void(base_size = 18) +
  coord_cartesian(expand = T, xlim = c(-1,1)) +
  theme(legend.position = "bottom") +
  labs(title = "Number of genes involved in distinctions with gain/loss of domains")

# rm(geneCounts_byClusters_tbl, geneCounts_gtbl)
```

# Another way to describe enrichments(except tiles)
```{r, fig.height=6, fig.width=8}
require(ggrepel)
theme_set(theme_classic(base_size = 25, base_family = "Arial"))
t_cells = setNames(pCells %>% group_by(characteristics..majorCluster) %>% summarise(cell_number = n()), c("cluster", "cell_number"))

t_numberbyCluster_tbl = left_join(enrichment_tbl %>% group_by(cluster) %>% summarise(gene_number = length(unique(gene_name))), t_cells, by = "cluster") 

p_vsCellnumber = t_numberbyCluster_tbl %>% 
  ggplot(aes(x = cell_number, y = gene_number)) +
  geom_point(size = 3) +
  geom_text_repel(aes(label = cluster), size = 6) +
  labs(x = "Cell number", y = "Number of enriched genes")

p_enrichedGenes = t_numberbyCluster_tbl %>% 
  ggplot(aes(x = cluster, y = gene_number)) +
  geom_col(size = 2) +
  geom_text(aes(label = gene_number, y = gene_number + 100), size = 6) +
  labs(x = "", y = "Number of enriched genes") +
  theme(axis.text.x = element_text(angle = 30, hjust = 1))
  

p_vsCellnumber
ggsave(plot = p_vsCellnumber, device = "pdf", width = 8, height = 6, filename = paste0( 
  WhereAmI("01.DATA/01.ASinLung/03.graphs/04.publishable/"), "EnrichedGenesbyClusterVsCellNumber.pdf"
    ))
p_enrichedGenes
ggsave(plot = p_enrichedGenes, device = "pdf", width = 8, height = 6, filename = paste0( 
  WhereAmI("01.DATA/01.ASinLung/03.graphs/04.publishable/"), "EnrichedGenesbyCluster.pdf"
    ))

```

## GO analysis for each cluster
```{r}
load("/media/jason/144E5C104E5BE94E/tmp/eGenesbyClusterCD8.RDa")
# require(clusterProfiler)
# t_eGenes_cluster = lapply(
#   pCells$characteristics..majorCluster %>% unique %>% sort,  
#   function(x){ with(enrichment_tbl, gene_name[cluster == x])}) %>% 
#   setNames(pCells$characteristics..majorCluster %>% unique %>% sort)
# 
# t_ego_cluster_lst = list()
# for(chn_cluster in names(t_eGenes_cluster)){
#   t_ego_cluster_lst[[chn_cluster]] = enrichGO(t_eGenes_cluster[[chn_cluster]] %>% XXXToEntrez(), OrgDb = "org.Hs.eg.db", ont="BP", readable=TRUE)
#   loginfo(chn_cluster, "done")
# }

# save(list = c("t_eGenes_cluster", "t_ego_cluster_lst", "ego_cluster_lst"), file = "/media/jason/144E5C104E5BE94E/tmp/eGenesbyClusterCD8.RDa")
# ego_cluster_lst = list()

# for(chn_cluster in setdiff(names(t_ego_cluster_lst), names(ego_cluster_lst))){
# # chn_cluster = names(t_ego_cluster_lst)[6]
#   ego_cluster_lst[[chn_cluster]] = clusterProfiler::simplify(t_ego_cluster_lst[[chn_cluster]])
#   cat(date(), chn_cluster, "done!\n")
#   gc()
# }

```

```{r, fig.width=12, fig.height=6}

topN = 10
for(x in names(ego_cluster_lst)){
  y = as_tibble(ego_cluster_lst[[x]])
  idx = order(y$p.adjust) %>% head(topN)
  y = y[idx, ]
  y$Description = factor(y$Description, levels = y$Description[order(y$p.adjust, decreasing = T)])
  print(
    y %>% 
      ggplot(aes(x = Description, y = -log10(p.adjust))) + 
      geom_col(fill = "red", width = 0.6) + 
      labs(title = x, x = NULL, y = "-log10(adjusted p value)") + 
      coord_flip() + 
      geom_text(aes(label = round(..y.., 2)), hjust = -0.4, size = 6) + 
      scale_y_continuous(position = "right", expand = expand_scale(mult = c(0, .5))) + 
      theme(axis.ticks.y = element_blank(), axis.line.y = element_blank())
  )
}

# mapply(ego_cluster_lst, FUN = function(x, title){dotplot(x, showCategory = 10) + labs(title = title)},  title = names(ego_cluster_lst), SIMPLIFY = F)
pdf(file = paste0(
      WhereAmI("01.DATA/01.ASinLung/03.graphs/04.publishable/"), "GObyCluster.pdf"
    ), title = chosen_gene, width = 12, height = 6)

for(x in names(ego_cluster_lst)){
  y = as_tibble(ego_cluster_lst[[x]])
  idx = order(y$p.adjust) %>% head(topN)
  y = y[idx, ]
  y$Description = factor(y$Description, levels = y$Description[order(y$p.adjust, decreasing = T)])
  print(
    y %>% 
      ggplot(aes(x = Description, y = -log10(p.adjust))) + 
      geom_col(fill = "red", width = 0.6) + 
      labs(title = x, x = NULL, y = "-log10(adjusted p value)") + 
      coord_flip() + 
      geom_text(aes(label = round(..y.., 2)), hjust = -0.4, size = 6) + 
      scale_y_continuous(position = "right", expand = expand_scale(mult = c(0, .5))) + 
      theme(axis.ticks.y = element_blank(), axis.line.y = element_blank())
  )
}

dev.off()

```

```{r, fig.height=8, fig.width=8}
# emapplot(ego_cluster_lst[[6]])
C6_enrichScores_tbl = enrichment_tbl %>% filter(cluster == "CD8_C6-LAYN") %>% group_by(gene_name) %>% dplyr::summarise(mean_enrich = mean(enrich_score)) %>% arrange(dplyr::desc(mean_enrich))

C6_enrichScores = setNames(C6_enrichScores_tbl$mean_enrich, nm = (C6_enrichScores_tbl$gene_name %>% XXXToEntrez()))

C6_gsea = gseGO(C6_enrichScores, ont = "BP", OrgDb = "org.Hs.eg.db", nPerm = 2000)


C6_gsea_tbl = as_tibble(C6_gsea)

C6_gsea_tbl %>% ggplot(aes(x = enrichmentScore, y = p.adjust)) +
  geom_point(aes(size = setSize))


idx = order(C6_gsea_tbl$p.adjust) %>% head(topN)
t_gsea_tbl = C6_gsea_tbl[idx, ]
t_gsea_tbl$Description = factor(t_gsea_tbl$Description, levels = t_gsea_tbl$Description[order(t_gsea_tbl$p.adjust, decreasing = T)])
t_gsea_tbl %>% 
  ggplot(aes(x = Description, y = enrichmentScore)) + 
  geom_col(fill = "red", width = 0.6) + 
  labs(title = "GSEA in C6", x = NULL) + 
  coord_flip() + 
  geom_text(aes(label = round(..y.., 2)), hjust = -0.4, size = 6) + 
  scale_y_continuous(position = "right", expand = expand_scale(mult = c(0, .5))) + 
  theme(axis.ticks.y = element_blank(), axis.line.y = element_blank())
  

openxlsx::write.xlsx(C6_gsea_tbl, file = "/home/jason/Storage/SYNC2/01.DATA/01.ASinLung/04.tables/GSEAwithGO_C6.xlsx", asTable = T)

chosen_descriptions = c(
  "response to hypoxia",
  "T cell activation",
  "cell activation",
  "T cell migration",
  "T cell receptor signaling pathway"
)

emapplot(C6_gsea, 20, color = "p.adjust")
for(chosen_description in chosen_descriptions){
  chosen_id = with(C6_gsea_tbl, ID[Description == chosen_description])
  p_gsea = gseaplot(C6_gsea, chosen_id, title = chosen_description)
  print(p_gsea)
  ggsave(plot = p_gsea, filename = paste0("/home/jason/Storage/SYNC2/01.DATA/01.ASinLung/03.graphs/04.publishable/enrichedInC6_gsea/", str_replace_all(chosen_description, "\\s", "_"), "_C6gsea.pdf"), device = "pdf", width = 8, height = 8)
}

```


# T cell subtypes vs. Usage distinctions

## Which T cell subtype tend to change from MIX 
```{r}
t_enrich_tbl = enrichment_filter_tbl %>% mutate(type = ifelse(grepl(id, pattern = "MIX"), "MIX", "isoform-dominated"))

chosen_clusters = t_enrich_tbl$cluster %>% unique %>% sort()

t_mat_lst = list()
for(chosen_cluster in chosen_clusters){
  t_mat = matrix(NA, 2, 2)
  colnames(t_mat) = c("MIX", "Isoform_dominated")
  rownames(t_mat) = c(chosen_cluster, "others")
  t_mat[1, 1] = (t_enrich_tbl %>% filter(cluster == chosen_cluster, type == "MIX"))$gene_name %>% unique() %>% length()
  t_mat[1, 2] = (t_enrich_tbl %>% filter(cluster == chosen_cluster, type != "MIX"))$gene_name %>% unique() %>% length()
  t_mat[2, 1] = (t_enrich_tbl %>% filter(cluster != chosen_cluster, type == "MIX"))$gene_name %>% unique() %>% length()
  t_mat[2, 2] = (t_enrich_tbl %>% filter(cluster != chosen_cluster, type != "MIX"))$gene_name %>% unique() %>% length()
  
  t_mat_lst[[chosen_cluster]] = t_mat
}

t_pval_lst = lapply(t_mat_lst, fisher.test)

t_pval_tbl = tibble(
  cluster = names(t_pval_lst), 
  OR = unlist(lapply(t_pval_lst, function(x) x$estimate)),
  p.value = unlist(lapply(t_pval_lst, function(x) x$p.value))
) %>% mutate(p.adjust = p.adjust(p.value), `-log10(p.adjust)` = -log10(p.adjust))

t_pval_tbl %>% 
  ggplot(aes(x = OR, y = `-log10(p.adjust)`)) + 
  geom_point(size = 4) + 
  geom_vline(xintercept = 1, linetype = 2) +
  geom_text_repel(aes(label = cluster), nudge_x = 0.1) +
  labs(x = "Odd ratio", title = "Enrichment of MIX genes in each cluster")

```

### show MIX genes in C6
```{r}
chosen_genes = (t_enrich_tbl %>% filter(cluster %>% grepl(pattern = "C6"), type == "MIX"))$gene_name %>% unique()

t_ego = enrichGO(chosen_genes %>% XXXToEntrez(), OrgDb = "org.Hs.eg.db", ont="BP", readable=TRUE)
t_ego2 = simplify(t_ego)
dotplot(t_ego2, showCategory = 10)
```